{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a53b3a87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# In case you do not have following packages installed, uncomment instalisation.\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "import functools\n",
    "from pathlib import Path\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#!pip install dask[complete];\n",
    "# you need to run these in case dask gives you error, it might need update.\n",
    "#!pip install --upgrade pandas \"dask[complete]\"\n",
    "#python -m pip install \"dask[dataframe]\" --upgrade\n",
    "#import dask.dataframe as dd\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "from sklearn.metrics import precision_score, recall_score, f1_score, roc_curve, accuracy_score\n",
    "from sklearn.metrics import roc_auc_score, confusion_matrix, precision_recall_curve\n",
    "\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.ensemble import GradientBoostingClassifier\n",
    "from sklearn.model_selection import RepeatedStratifiedKFold\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.inspection import permutation_importance       \n",
    "\n",
    "from sklearn.feature_selection import SelectFromModel\n",
    "from sklearn.utils import class_weight\n",
    "\n",
    "#!pip3 install xgboost\n",
    "#from xgboost import XGBClassifier\n",
    "\n",
    "#!pip install conorm\n",
    "#import conorm # for tmm normalisation\n",
    "\n",
    "#!pip3 install pydeseq2 or pip install pydeseq2\n",
    "#from pydeseq2.dds import DeseqDataSet\n",
    "#from pydeseq2.ds import DeseqStats\n",
    "#from pydeseq2.utils import load_example_data\n",
    "\n",
    "\n",
    "\n",
    "#to install R :\n",
    "#conda install -c r r-irkernel\n",
    "\n",
    "#to install a library from R\n",
    "#!pip install library edgeR\n",
    "# pip install rpy2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6436b65",
   "metadata": {},
   "source": [
    "### Data for Imaging of Parkinson ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3c33c11d",
   "metadata": {},
   "outputs": [],
   "source": [
    "path3 = Path(\"/home/znazari/data/open_proteomic/\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "fafcc486",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene stable ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT-TF</td>\n",
       "      <td>ENSG00000210049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>ENSG00000211459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT-TV</td>\n",
       "      <td>ENSG00000210077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>ENSG00000210082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>ENSG00000209082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274613</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000284882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274614</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000284882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274615</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000289881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274616</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000289881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274617</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000285325</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>274618 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene name   Gene stable ID\n",
       "0          MT-TF  ENSG00000210049\n",
       "1        MT-RNR1  ENSG00000211459\n",
       "2          MT-TV  ENSG00000210077\n",
       "3        MT-RNR2  ENSG00000210082\n",
       "4         MT-TL1  ENSG00000209082\n",
       "...          ...              ...\n",
       "274613       NaN  ENSG00000284882\n",
       "274614       NaN  ENSG00000284882\n",
       "274615       NaN  ENSG00000289881\n",
       "274616       NaN  ENSG00000289881\n",
       "274617       NaN  ENSG00000285325\n",
       "\n",
       "[274618 rows x 2 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensemble = pd.read_csv(path3/'Ensemble_GeneID_TranscriptID_Symbol_mart_export.txt',delimiter='\\t')\n",
    "gin = ensemble[['Gene name','Gene stable ID']]\n",
    "gin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "de5ba5aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Transcript name</th>\n",
       "      <th>HGNC symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>MT-TF-201</td>\n",
       "      <td>MT-TF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>ENST00000389680.2</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>MT-RNR1-201</td>\n",
       "      <td>MT-RNR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>ENST00000387342.1</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>MT-TV-201</td>\n",
       "      <td>MT-TV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>ENST00000387347.2</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>MT-RNR2-201</td>\n",
       "      <td>MT-RNR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>ENST00000386347.1</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>MT-TL1-201</td>\n",
       "      <td>MT-TL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274613</th>\n",
       "      <td>ENSG00000284882</td>\n",
       "      <td>ENSG00000284882.2</td>\n",
       "      <td>ENST00000646567</td>\n",
       "      <td>ENST00000646567.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274614</th>\n",
       "      <td>ENSG00000284882</td>\n",
       "      <td>ENSG00000284882.2</td>\n",
       "      <td>ENST00000644618</td>\n",
       "      <td>ENST00000644618.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274615</th>\n",
       "      <td>ENSG00000289881</td>\n",
       "      <td>ENSG00000289881.1</td>\n",
       "      <td>ENST00000701339</td>\n",
       "      <td>ENST00000701339.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274616</th>\n",
       "      <td>ENSG00000289881</td>\n",
       "      <td>ENSG00000289881.1</td>\n",
       "      <td>ENST00000701824</td>\n",
       "      <td>ENST00000701824.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274617</th>\n",
       "      <td>ENSG00000285325</td>\n",
       "      <td>ENSG00000285325.1</td>\n",
       "      <td>ENST00000643511</td>\n",
       "      <td>ENST00000643511.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>274618 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0       ENSG00000210049      ENSG00000210049.1      ENST00000387314   \n",
       "1       ENSG00000211459      ENSG00000211459.2      ENST00000389680   \n",
       "2       ENSG00000210077      ENSG00000210077.1      ENST00000387342   \n",
       "3       ENSG00000210082      ENSG00000210082.2      ENST00000387347   \n",
       "4       ENSG00000209082      ENSG00000209082.1      ENST00000386347   \n",
       "...                 ...                    ...                  ...   \n",
       "274613  ENSG00000284882      ENSG00000284882.2      ENST00000646567   \n",
       "274614  ENSG00000284882      ENSG00000284882.2      ENST00000644618   \n",
       "274615  ENSG00000289881      ENSG00000289881.1      ENST00000701339   \n",
       "274616  ENSG00000289881      ENSG00000289881.1      ENST00000701824   \n",
       "274617  ENSG00000285325      ENSG00000285325.1      ENST00000643511   \n",
       "\n",
       "       Transcript stable ID version Gene name Transcript name HGNC symbol  \n",
       "0                 ENST00000387314.1     MT-TF       MT-TF-201       MT-TF  \n",
       "1                 ENST00000389680.2   MT-RNR1     MT-RNR1-201     MT-RNR1  \n",
       "2                 ENST00000387342.1     MT-TV       MT-TV-201       MT-TV  \n",
       "3                 ENST00000387347.2   MT-RNR2     MT-RNR2-201     MT-RNR2  \n",
       "4                 ENST00000386347.1    MT-TL1      MT-TL1-201      MT-TL1  \n",
       "...                             ...       ...             ...         ...  \n",
       "274613            ENST00000646567.2       NaN             NaN         NaN  \n",
       "274614            ENST00000644618.1       NaN             NaN         NaN  \n",
       "274615            ENST00000701339.1       NaN             NaN         NaN  \n",
       "274616            ENST00000701824.1       NaN             NaN         NaN  \n",
       "274617            ENST00000643511.1       NaN             NaN         NaN  \n",
       "\n",
       "[274618 rows x 7 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensemble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f295bb12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SOMA_SEQ_ID</th>\n",
       "      <th>TARGET_GENE_SYMBOL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10000-28_3</td>\n",
       "      <td>CRYBB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10001-7_3</td>\n",
       "      <td>RAF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10003-15_3</td>\n",
       "      <td>ZNF41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10006-25_3</td>\n",
       "      <td>ELK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10008-43_3</td>\n",
       "      <td>GUCA1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5188</th>\n",
       "      <td>9993-11_3</td>\n",
       "      <td>ZNF264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5189</th>\n",
       "      <td>9994-217_3</td>\n",
       "      <td>ATP4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5190</th>\n",
       "      <td>9995-6_3</td>\n",
       "      <td>DUT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5191</th>\n",
       "      <td>9997-12_3</td>\n",
       "      <td>UBXN4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5192</th>\n",
       "      <td>9999-1_3</td>\n",
       "      <td>IRF6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5193 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SOMA_SEQ_ID TARGET_GENE_SYMBOL\n",
       "0     10000-28_3             CRYBB2\n",
       "1      10001-7_3               RAF1\n",
       "2     10003-15_3              ZNF41\n",
       "3     10006-25_3               ELK1\n",
       "4     10008-43_3             GUCA1A\n",
       "...          ...                ...\n",
       "5188   9993-11_3             ZNF264\n",
       "5189  9994-217_3              ATP4B\n",
       "5190    9995-6_3                DUT\n",
       "5191   9997-12_3              UBXN4\n",
       "5192    9999-1_3               IRF6\n",
       "\n",
       "[5193 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteomic_annotation = pd.read_csv(path3/\"PPMI_Project_151_pqtl_Analysis_Annotations_20210210.csv\",delimiter=',')\n",
    "symb = proteomic_annotation[['SOMA_SEQ_ID','TARGET_GENE_SYMBOL']]\n",
    "symb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "214726e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the base file name and path\n",
    "base_file_name = \"Project_151_pQTL_in_CSF_{}_of_7_Batch_Corrected_.csv\"\n",
    "\n",
    "# Number of files\n",
    "num_files = 7\n",
    "\n",
    "# List to store DataFrames\n",
    "dfs = []\n",
    "\n",
    "# Loop through the file indices and read each file\n",
    "for file_index in range(1, num_files + 1):\n",
    "    file_name = base_file_name.format(file_index)\n",
    "    file_path = path3 / file_name\n",
    "    \n",
    "    # Check if the file exists before attempting to read it\n",
    "    if file_path.is_file():\n",
    "        # Read the CSV file and append it to the list\n",
    "        df = pd.read_csv(file_path, delimiter=',')\n",
    "        dfs.append(df)\n",
    "    else:\n",
    "        print(f\"File {file_name} not found.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "76823aba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Concatenate all DataFrames into a single DataFrame\n",
    "result_df = pd.concat(dfs, ignore_index=True)\n",
    "\n",
    "# Filter out patients diagnosed as Prodromal\n",
    "result_df = result_df[result_df['COHORT'] != 'Prodromal']\n",
    "\n",
    "# Extract the two columns TESTNAME and PATNO\n",
    "result_subset = result_df[['TESTNAME', 'PATNO']]\n",
    "\n",
    "# Pivot the DataFrame to get the desired format\n",
    "#result_pivot = result_subset.pivot_table(index='TESTNAME', columns='PATNO', aggfunc=len, fill_value=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e21b559e",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_result_subset = result_subset.drop_duplicates(subset=['TESTNAME'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "91280ae6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4785, 2)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#IMPORTANT TESTNAME: are the main names associated with CSF:\n",
    "unique_result_subset.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e931b71e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "679827ca",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "76527e9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_values_symb = symb[symb['SOMA_SEQ_ID'].isin(unique_result_subset['TESTNAME'])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d27af524",
   "metadata": {},
   "outputs": [],
   "source": [
    "symb_gene= common_values_symb.drop_duplicates(subset=['SOMA_SEQ_ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "32b7a8ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "symb_genesymb= symb_gene.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b564519f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SOMA_SEQ_ID</th>\n",
       "      <th>TARGET_GENE_SYMBOL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10000-28_3</td>\n",
       "      <td>CRYBB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10001-7_3</td>\n",
       "      <td>RAF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10003-15_3</td>\n",
       "      <td>ZNF41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10006-25_3</td>\n",
       "      <td>ELK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10008-43_3</td>\n",
       "      <td>GUCA1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4780</th>\n",
       "      <td>9993-11_3</td>\n",
       "      <td>ZNF264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4781</th>\n",
       "      <td>9994-217_3</td>\n",
       "      <td>ATP4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4782</th>\n",
       "      <td>9995-6_3</td>\n",
       "      <td>DUT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4783</th>\n",
       "      <td>9997-12_3</td>\n",
       "      <td>UBXN4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4784</th>\n",
       "      <td>9999-1_3</td>\n",
       "      <td>IRF6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4785 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SOMA_SEQ_ID TARGET_GENE_SYMBOL\n",
       "0     10000-28_3             CRYBB2\n",
       "1      10001-7_3               RAF1\n",
       "2     10003-15_3              ZNF41\n",
       "3     10006-25_3               ELK1\n",
       "4     10008-43_3             GUCA1A\n",
       "...          ...                ...\n",
       "4780   9993-11_3             ZNF264\n",
       "4781  9994-217_3              ATP4B\n",
       "4782    9995-6_3                DUT\n",
       "4783   9997-12_3              UBXN4\n",
       "4784    9999-1_3               IRF6\n",
       "\n",
       "[4785 rows x 2 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "symb_genesymb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "48230b1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_ensmble_genesyb = gin[gin['Gene name'].isin(symb_genesymb['TARGET_GENE_SYMBOL'])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "542f3d5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(42633, 2)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ensmble_genesyb.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "1a3281d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_ensg = gene_ensmble_genesyb.drop_duplicates(subset=['Gene name']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "9ca57ac8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene stable ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>INTS3</td>\n",
       "      <td>ENSG00000262826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KIR3DL2</td>\n",
       "      <td>ENSG00000273735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KIR3DL3</td>\n",
       "      <td>ENSG00000276433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PTCHD3</td>\n",
       "      <td>ENSG00000276595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KIR2DL4</td>\n",
       "      <td>ENSG00000276779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3996</th>\n",
       "      <td>BTG2</td>\n",
       "      <td>ENSG00000159388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3997</th>\n",
       "      <td>KCNAB2</td>\n",
       "      <td>ENSG00000069424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3998</th>\n",
       "      <td>NUP210L</td>\n",
       "      <td>ENSG00000143552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3999</th>\n",
       "      <td>NTRK1</td>\n",
       "      <td>ENSG00000198400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4000</th>\n",
       "      <td>ACOT7</td>\n",
       "      <td>ENSG00000097021</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4001 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene name   Gene stable ID\n",
       "0        INTS3  ENSG00000262826\n",
       "1      KIR3DL2  ENSG00000273735\n",
       "2      KIR3DL3  ENSG00000276433\n",
       "3       PTCHD3  ENSG00000276595\n",
       "4      KIR2DL4  ENSG00000276779\n",
       "...        ...              ...\n",
       "3996      BTG2  ENSG00000159388\n",
       "3997    KCNAB2  ENSG00000069424\n",
       "3998   NUP210L  ENSG00000143552\n",
       "3999     NTRK1  ENSG00000198400\n",
       "4000     ACOT7  ENSG00000097021\n",
       "\n",
       "[4001 rows x 2 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_ensg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dac8bd5c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "7a3fc6a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading the already preprocessed data:\n",
    "\n",
    "path2 = Path(\"/home/znazari/data\") # where the output data will be saved at the end.\n",
    "\n",
    "# with all the filtered genes:\n",
    "\n",
    "ir3_rna_step_vst =  pd.read_csv(path2/'mydata_Log_CPM_filtered_bact_sex_effect_removed_RIN_covariate.txt',delimiter='\\t' )\n",
    "\n",
    "diagnosis = pd.read_csv(path2/'patients_HC_PK_diagnosis.csv')\n",
    "\n",
    "# mapping diagnosis to zero and one.\n",
    "diagnosis['COHORT_DEFINITION'] = diagnosis['COHORT_DEFINITION'].map({'Healthy Control': 0, \"Parkinson's Disease\": 1})\n",
    "\n",
    "# X: feature matrix, y: the target variable\n",
    "\n",
    "X=ir3_rna_step_vst.T\n",
    "y=diagnosis['COHORT_DEFINITION']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "70ca9253",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3000</th>\n",
       "      <th>3001</th>\n",
       "      <th>3002</th>\n",
       "      <th>3003</th>\n",
       "      <th>3004</th>\n",
       "      <th>3006</th>\n",
       "      <th>3007</th>\n",
       "      <th>3008</th>\n",
       "      <th>3009</th>\n",
       "      <th>3010</th>\n",
       "      <th>...</th>\n",
       "      <th>4121</th>\n",
       "      <th>4122</th>\n",
       "      <th>4123</th>\n",
       "      <th>4124</th>\n",
       "      <th>4125</th>\n",
       "      <th>4126</th>\n",
       "      <th>4135</th>\n",
       "      <th>4136</th>\n",
       "      <th>4139</th>\n",
       "      <th>41410</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>3.845119</td>\n",
       "      <td>5.213262</td>\n",
       "      <td>4.837730</td>\n",
       "      <td>5.101270</td>\n",
       "      <td>4.974751</td>\n",
       "      <td>4.825708</td>\n",
       "      <td>5.011214</td>\n",
       "      <td>5.229956</td>\n",
       "      <td>5.174531</td>\n",
       "      <td>5.284571</td>\n",
       "      <td>...</td>\n",
       "      <td>4.451699</td>\n",
       "      <td>5.408345</td>\n",
       "      <td>5.247058</td>\n",
       "      <td>4.584573</td>\n",
       "      <td>4.398598</td>\n",
       "      <td>5.449524</td>\n",
       "      <td>5.015613</td>\n",
       "      <td>3.987076</td>\n",
       "      <td>5.496642</td>\n",
       "      <td>4.963686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>5.648852</td>\n",
       "      <td>6.202708</td>\n",
       "      <td>5.624314</td>\n",
       "      <td>6.067816</td>\n",
       "      <td>6.092646</td>\n",
       "      <td>6.223112</td>\n",
       "      <td>6.003440</td>\n",
       "      <td>6.010872</td>\n",
       "      <td>6.272480</td>\n",
       "      <td>6.285266</td>\n",
       "      <td>...</td>\n",
       "      <td>5.863215</td>\n",
       "      <td>6.422896</td>\n",
       "      <td>6.368654</td>\n",
       "      <td>6.114092</td>\n",
       "      <td>5.798945</td>\n",
       "      <td>6.383039</td>\n",
       "      <td>6.290690</td>\n",
       "      <td>5.179048</td>\n",
       "      <td>6.293745</td>\n",
       "      <td>6.058590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>3.817007</td>\n",
       "      <td>4.095830</td>\n",
       "      <td>4.012751</td>\n",
       "      <td>4.262379</td>\n",
       "      <td>3.969580</td>\n",
       "      <td>4.374808</td>\n",
       "      <td>4.417553</td>\n",
       "      <td>4.331596</td>\n",
       "      <td>4.675491</td>\n",
       "      <td>4.348990</td>\n",
       "      <td>...</td>\n",
       "      <td>3.957820</td>\n",
       "      <td>4.518421</td>\n",
       "      <td>4.424952</td>\n",
       "      <td>4.145217</td>\n",
       "      <td>4.194468</td>\n",
       "      <td>4.628795</td>\n",
       "      <td>4.747158</td>\n",
       "      <td>3.569694</td>\n",
       "      <td>4.025577</td>\n",
       "      <td>4.268701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>8.463275</td>\n",
       "      <td>8.810743</td>\n",
       "      <td>8.368607</td>\n",
       "      <td>8.466923</td>\n",
       "      <td>8.927912</td>\n",
       "      <td>9.043402</td>\n",
       "      <td>8.681277</td>\n",
       "      <td>8.235545</td>\n",
       "      <td>8.067889</td>\n",
       "      <td>8.631065</td>\n",
       "      <td>...</td>\n",
       "      <td>8.416943</td>\n",
       "      <td>9.104202</td>\n",
       "      <td>8.838484</td>\n",
       "      <td>9.193292</td>\n",
       "      <td>8.217039</td>\n",
       "      <td>8.772340</td>\n",
       "      <td>8.497247</td>\n",
       "      <td>8.612951</td>\n",
       "      <td>8.695999</td>\n",
       "      <td>8.676493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <td>1.944928</td>\n",
       "      <td>3.384911</td>\n",
       "      <td>2.522713</td>\n",
       "      <td>4.085667</td>\n",
       "      <td>3.693875</td>\n",
       "      <td>2.544924</td>\n",
       "      <td>3.968554</td>\n",
       "      <td>4.085922</td>\n",
       "      <td>4.811027</td>\n",
       "      <td>4.093739</td>\n",
       "      <td>...</td>\n",
       "      <td>2.209538</td>\n",
       "      <td>2.390307</td>\n",
       "      <td>3.974330</td>\n",
       "      <td>2.196031</td>\n",
       "      <td>2.196303</td>\n",
       "      <td>3.910535</td>\n",
       "      <td>4.403068</td>\n",
       "      <td>2.359305</td>\n",
       "      <td>2.445675</td>\n",
       "      <td>3.267041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285219</th>\n",
       "      <td>2.721748</td>\n",
       "      <td>1.638440</td>\n",
       "      <td>3.771521</td>\n",
       "      <td>1.812421</td>\n",
       "      <td>2.549122</td>\n",
       "      <td>2.058582</td>\n",
       "      <td>2.125234</td>\n",
       "      <td>1.765596</td>\n",
       "      <td>2.069582</td>\n",
       "      <td>1.764143</td>\n",
       "      <td>...</td>\n",
       "      <td>2.495753</td>\n",
       "      <td>1.830501</td>\n",
       "      <td>1.218611</td>\n",
       "      <td>2.165035</td>\n",
       "      <td>3.126425</td>\n",
       "      <td>1.943526</td>\n",
       "      <td>0.648852</td>\n",
       "      <td>2.959793</td>\n",
       "      <td>2.179546</td>\n",
       "      <td>2.136740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285230</th>\n",
       "      <td>2.322872</td>\n",
       "      <td>2.517009</td>\n",
       "      <td>2.599599</td>\n",
       "      <td>2.386956</td>\n",
       "      <td>2.851771</td>\n",
       "      <td>2.737816</td>\n",
       "      <td>2.894924</td>\n",
       "      <td>2.388526</td>\n",
       "      <td>2.458358</td>\n",
       "      <td>2.695723</td>\n",
       "      <td>...</td>\n",
       "      <td>2.169890</td>\n",
       "      <td>2.481177</td>\n",
       "      <td>2.779666</td>\n",
       "      <td>2.643108</td>\n",
       "      <td>2.415556</td>\n",
       "      <td>2.438769</td>\n",
       "      <td>2.464313</td>\n",
       "      <td>2.474065</td>\n",
       "      <td>2.487454</td>\n",
       "      <td>2.544368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285280</th>\n",
       "      <td>4.541140</td>\n",
       "      <td>1.709681</td>\n",
       "      <td>2.196385</td>\n",
       "      <td>2.727775</td>\n",
       "      <td>2.760335</td>\n",
       "      <td>2.204436</td>\n",
       "      <td>2.655007</td>\n",
       "      <td>2.044956</td>\n",
       "      <td>1.865374</td>\n",
       "      <td>1.846651</td>\n",
       "      <td>...</td>\n",
       "      <td>2.344181</td>\n",
       "      <td>2.573584</td>\n",
       "      <td>2.253876</td>\n",
       "      <td>2.459218</td>\n",
       "      <td>3.058924</td>\n",
       "      <td>2.916741</td>\n",
       "      <td>2.501802</td>\n",
       "      <td>2.977444</td>\n",
       "      <td>2.196541</td>\n",
       "      <td>2.727432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285399</th>\n",
       "      <td>5.099339</td>\n",
       "      <td>5.353763</td>\n",
       "      <td>4.708896</td>\n",
       "      <td>5.048748</td>\n",
       "      <td>5.349433</td>\n",
       "      <td>5.567314</td>\n",
       "      <td>5.192088</td>\n",
       "      <td>5.066864</td>\n",
       "      <td>5.211345</td>\n",
       "      <td>5.326240</td>\n",
       "      <td>...</td>\n",
       "      <td>5.208116</td>\n",
       "      <td>5.197208</td>\n",
       "      <td>5.421681</td>\n",
       "      <td>5.656614</td>\n",
       "      <td>5.370812</td>\n",
       "      <td>5.384910</td>\n",
       "      <td>5.832674</td>\n",
       "      <td>4.774830</td>\n",
       "      <td>4.800063</td>\n",
       "      <td>5.287198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285410</th>\n",
       "      <td>5.241609</td>\n",
       "      <td>6.019862</td>\n",
       "      <td>5.414418</td>\n",
       "      <td>5.555605</td>\n",
       "      <td>5.554071</td>\n",
       "      <td>5.654699</td>\n",
       "      <td>5.836791</td>\n",
       "      <td>5.971206</td>\n",
       "      <td>6.017686</td>\n",
       "      <td>6.123887</td>\n",
       "      <td>...</td>\n",
       "      <td>5.526208</td>\n",
       "      <td>6.105318</td>\n",
       "      <td>5.998763</td>\n",
       "      <td>5.768046</td>\n",
       "      <td>5.472123</td>\n",
       "      <td>5.950408</td>\n",
       "      <td>5.878510</td>\n",
       "      <td>5.187680</td>\n",
       "      <td>5.677015</td>\n",
       "      <td>5.755279</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13104 rows × 583 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     3000      3001      3002      3003      3004      3006  \\\n",
       "ENSG00000000419  3.845119  5.213262  4.837730  5.101270  4.974751  4.825708   \n",
       "ENSG00000000457  5.648852  6.202708  5.624314  6.067816  6.092646  6.223112   \n",
       "ENSG00000000460  3.817007  4.095830  4.012751  4.262379  3.969580  4.374808   \n",
       "ENSG00000000938  8.463275  8.810743  8.368607  8.466923  8.927912  9.043402   \n",
       "ENSG00000000971  1.944928  3.384911  2.522713  4.085667  3.693875  2.544924   \n",
       "...                   ...       ...       ...       ...       ...       ...   \n",
       "ENSG00000285219  2.721748  1.638440  3.771521  1.812421  2.549122  2.058582   \n",
       "ENSG00000285230  2.322872  2.517009  2.599599  2.386956  2.851771  2.737816   \n",
       "ENSG00000285280  4.541140  1.709681  2.196385  2.727775  2.760335  2.204436   \n",
       "ENSG00000285399  5.099339  5.353763  4.708896  5.048748  5.349433  5.567314   \n",
       "ENSG00000285410  5.241609  6.019862  5.414418  5.555605  5.554071  5.654699   \n",
       "\n",
       "                     3007      3008      3009      3010  ...      4121  \\\n",
       "ENSG00000000419  5.011214  5.229956  5.174531  5.284571  ...  4.451699   \n",
       "ENSG00000000457  6.003440  6.010872  6.272480  6.285266  ...  5.863215   \n",
       "ENSG00000000460  4.417553  4.331596  4.675491  4.348990  ...  3.957820   \n",
       "ENSG00000000938  8.681277  8.235545  8.067889  8.631065  ...  8.416943   \n",
       "ENSG00000000971  3.968554  4.085922  4.811027  4.093739  ...  2.209538   \n",
       "...                   ...       ...       ...       ...  ...       ...   \n",
       "ENSG00000285219  2.125234  1.765596  2.069582  1.764143  ...  2.495753   \n",
       "ENSG00000285230  2.894924  2.388526  2.458358  2.695723  ...  2.169890   \n",
       "ENSG00000285280  2.655007  2.044956  1.865374  1.846651  ...  2.344181   \n",
       "ENSG00000285399  5.192088  5.066864  5.211345  5.326240  ...  5.208116   \n",
       "ENSG00000285410  5.836791  5.971206  6.017686  6.123887  ...  5.526208   \n",
       "\n",
       "                     4122      4123      4124      4125      4126      4135  \\\n",
       "ENSG00000000419  5.408345  5.247058  4.584573  4.398598  5.449524  5.015613   \n",
       "ENSG00000000457  6.422896  6.368654  6.114092  5.798945  6.383039  6.290690   \n",
       "ENSG00000000460  4.518421  4.424952  4.145217  4.194468  4.628795  4.747158   \n",
       "ENSG00000000938  9.104202  8.838484  9.193292  8.217039  8.772340  8.497247   \n",
       "ENSG00000000971  2.390307  3.974330  2.196031  2.196303  3.910535  4.403068   \n",
       "...                   ...       ...       ...       ...       ...       ...   \n",
       "ENSG00000285219  1.830501  1.218611  2.165035  3.126425  1.943526  0.648852   \n",
       "ENSG00000285230  2.481177  2.779666  2.643108  2.415556  2.438769  2.464313   \n",
       "ENSG00000285280  2.573584  2.253876  2.459218  3.058924  2.916741  2.501802   \n",
       "ENSG00000285399  5.197208  5.421681  5.656614  5.370812  5.384910  5.832674   \n",
       "ENSG00000285410  6.105318  5.998763  5.768046  5.472123  5.950408  5.878510   \n",
       "\n",
       "                     4136      4139     41410  \n",
       "ENSG00000000419  3.987076  5.496642  4.963686  \n",
       "ENSG00000000457  5.179048  6.293745  6.058590  \n",
       "ENSG00000000460  3.569694  4.025577  4.268701  \n",
       "ENSG00000000938  8.612951  8.695999  8.676493  \n",
       "ENSG00000000971  2.359305  2.445675  3.267041  \n",
       "...                   ...       ...       ...  \n",
       "ENSG00000285219  2.959793  2.179546  2.136740  \n",
       "ENSG00000285230  2.474065  2.487454  2.544368  \n",
       "ENSG00000285280  2.977444  2.196541  2.727432  \n",
       "ENSG00000285399  4.774830  4.800063  5.287198  \n",
       "ENSG00000285410  5.187680  5.677015  5.755279  \n",
       "\n",
       "[13104 rows x 583 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ir3_rna_step_vst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92b7626c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "500655b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_df = ir3_rna_step_vst[ir3_rna_step_vst.index.isin(unique_ensg['Gene stable ID'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "07e3977f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3000</th>\n",
       "      <th>3001</th>\n",
       "      <th>3002</th>\n",
       "      <th>3003</th>\n",
       "      <th>3004</th>\n",
       "      <th>3006</th>\n",
       "      <th>3007</th>\n",
       "      <th>3008</th>\n",
       "      <th>3009</th>\n",
       "      <th>3010</th>\n",
       "      <th>...</th>\n",
       "      <th>4121</th>\n",
       "      <th>4122</th>\n",
       "      <th>4123</th>\n",
       "      <th>4124</th>\n",
       "      <th>4125</th>\n",
       "      <th>4126</th>\n",
       "      <th>4135</th>\n",
       "      <th>4136</th>\n",
       "      <th>4139</th>\n",
       "      <th>41410</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>8.463275</td>\n",
       "      <td>8.810743</td>\n",
       "      <td>8.368607</td>\n",
       "      <td>8.466923</td>\n",
       "      <td>8.927912</td>\n",
       "      <td>9.043402</td>\n",
       "      <td>8.681277</td>\n",
       "      <td>8.235545</td>\n",
       "      <td>8.067889</td>\n",
       "      <td>8.631065</td>\n",
       "      <td>...</td>\n",
       "      <td>8.416943</td>\n",
       "      <td>9.104202</td>\n",
       "      <td>8.838484</td>\n",
       "      <td>9.193292</td>\n",
       "      <td>8.217039</td>\n",
       "      <td>8.772340</td>\n",
       "      <td>8.497247</td>\n",
       "      <td>8.612951</td>\n",
       "      <td>8.695999</td>\n",
       "      <td>8.676493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <td>1.944928</td>\n",
       "      <td>3.384911</td>\n",
       "      <td>2.522713</td>\n",
       "      <td>4.085667</td>\n",
       "      <td>3.693875</td>\n",
       "      <td>2.544924</td>\n",
       "      <td>3.968554</td>\n",
       "      <td>4.085922</td>\n",
       "      <td>4.811027</td>\n",
       "      <td>4.093739</td>\n",
       "      <td>...</td>\n",
       "      <td>2.209538</td>\n",
       "      <td>2.390307</td>\n",
       "      <td>3.974330</td>\n",
       "      <td>2.196031</td>\n",
       "      <td>2.196303</td>\n",
       "      <td>3.910535</td>\n",
       "      <td>4.403068</td>\n",
       "      <td>2.359305</td>\n",
       "      <td>2.445675</td>\n",
       "      <td>3.267041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000002330</th>\n",
       "      <td>3.571424</td>\n",
       "      <td>3.959793</td>\n",
       "      <td>3.388754</td>\n",
       "      <td>4.085483</td>\n",
       "      <td>3.756835</td>\n",
       "      <td>3.986339</td>\n",
       "      <td>3.665170</td>\n",
       "      <td>3.716128</td>\n",
       "      <td>4.060661</td>\n",
       "      <td>4.147625</td>\n",
       "      <td>...</td>\n",
       "      <td>3.368043</td>\n",
       "      <td>4.429167</td>\n",
       "      <td>3.945994</td>\n",
       "      <td>4.056662</td>\n",
       "      <td>2.965991</td>\n",
       "      <td>4.581369</td>\n",
       "      <td>4.673969</td>\n",
       "      <td>3.146942</td>\n",
       "      <td>4.272373</td>\n",
       "      <td>3.885313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000002549</th>\n",
       "      <td>5.872790</td>\n",
       "      <td>5.908892</td>\n",
       "      <td>6.004722</td>\n",
       "      <td>6.073372</td>\n",
       "      <td>5.541370</td>\n",
       "      <td>7.682882</td>\n",
       "      <td>5.899073</td>\n",
       "      <td>5.657070</td>\n",
       "      <td>6.353882</td>\n",
       "      <td>5.949753</td>\n",
       "      <td>...</td>\n",
       "      <td>6.898440</td>\n",
       "      <td>6.502385</td>\n",
       "      <td>6.639824</td>\n",
       "      <td>5.425128</td>\n",
       "      <td>5.561972</td>\n",
       "      <td>6.668685</td>\n",
       "      <td>7.166978</td>\n",
       "      <td>6.111543</td>\n",
       "      <td>6.510586</td>\n",
       "      <td>6.325302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000002587</th>\n",
       "      <td>0.139701</td>\n",
       "      <td>1.312890</td>\n",
       "      <td>1.736479</td>\n",
       "      <td>0.976318</td>\n",
       "      <td>1.218935</td>\n",
       "      <td>0.709119</td>\n",
       "      <td>1.053174</td>\n",
       "      <td>1.422313</td>\n",
       "      <td>2.073881</td>\n",
       "      <td>1.777967</td>\n",
       "      <td>...</td>\n",
       "      <td>1.751712</td>\n",
       "      <td>0.780928</td>\n",
       "      <td>1.562588</td>\n",
       "      <td>0.983353</td>\n",
       "      <td>1.199543</td>\n",
       "      <td>0.721738</td>\n",
       "      <td>-0.725075</td>\n",
       "      <td>1.494510</td>\n",
       "      <td>1.673137</td>\n",
       "      <td>0.951384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271303</th>\n",
       "      <td>4.141592</td>\n",
       "      <td>5.154501</td>\n",
       "      <td>3.158052</td>\n",
       "      <td>5.093705</td>\n",
       "      <td>3.283577</td>\n",
       "      <td>5.812244</td>\n",
       "      <td>4.497693</td>\n",
       "      <td>4.390843</td>\n",
       "      <td>3.943871</td>\n",
       "      <td>4.750242</td>\n",
       "      <td>...</td>\n",
       "      <td>2.731588</td>\n",
       "      <td>4.644567</td>\n",
       "      <td>4.741196</td>\n",
       "      <td>4.482349</td>\n",
       "      <td>3.986768</td>\n",
       "      <td>5.829626</td>\n",
       "      <td>5.499389</td>\n",
       "      <td>3.893179</td>\n",
       "      <td>5.252354</td>\n",
       "      <td>4.670282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271605</th>\n",
       "      <td>3.253041</td>\n",
       "      <td>3.349505</td>\n",
       "      <td>4.194189</td>\n",
       "      <td>4.647757</td>\n",
       "      <td>3.770106</td>\n",
       "      <td>4.032242</td>\n",
       "      <td>3.725165</td>\n",
       "      <td>4.014289</td>\n",
       "      <td>3.882068</td>\n",
       "      <td>3.648119</td>\n",
       "      <td>...</td>\n",
       "      <td>4.301260</td>\n",
       "      <td>3.829374</td>\n",
       "      <td>4.916163</td>\n",
       "      <td>4.849648</td>\n",
       "      <td>3.918818</td>\n",
       "      <td>3.755242</td>\n",
       "      <td>4.052355</td>\n",
       "      <td>3.537703</td>\n",
       "      <td>4.062805</td>\n",
       "      <td>3.979138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000272325</th>\n",
       "      <td>6.689842</td>\n",
       "      <td>6.335426</td>\n",
       "      <td>6.208001</td>\n",
       "      <td>6.495175</td>\n",
       "      <td>6.384606</td>\n",
       "      <td>6.810942</td>\n",
       "      <td>6.646111</td>\n",
       "      <td>6.402878</td>\n",
       "      <td>6.308627</td>\n",
       "      <td>6.737337</td>\n",
       "      <td>...</td>\n",
       "      <td>6.560920</td>\n",
       "      <td>6.703202</td>\n",
       "      <td>6.730188</td>\n",
       "      <td>6.621639</td>\n",
       "      <td>6.583015</td>\n",
       "      <td>6.837237</td>\n",
       "      <td>6.500973</td>\n",
       "      <td>6.102476</td>\n",
       "      <td>6.663363</td>\n",
       "      <td>6.564332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000272886</th>\n",
       "      <td>5.031533</td>\n",
       "      <td>5.461262</td>\n",
       "      <td>4.897880</td>\n",
       "      <td>5.414562</td>\n",
       "      <td>5.042777</td>\n",
       "      <td>5.492468</td>\n",
       "      <td>5.444711</td>\n",
       "      <td>5.371691</td>\n",
       "      <td>5.520976</td>\n",
       "      <td>5.315246</td>\n",
       "      <td>...</td>\n",
       "      <td>5.277152</td>\n",
       "      <td>5.400093</td>\n",
       "      <td>5.523296</td>\n",
       "      <td>5.426866</td>\n",
       "      <td>5.269029</td>\n",
       "      <td>5.388664</td>\n",
       "      <td>5.499137</td>\n",
       "      <td>4.822674</td>\n",
       "      <td>5.094705</td>\n",
       "      <td>5.293258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277586</th>\n",
       "      <td>0.955260</td>\n",
       "      <td>1.599759</td>\n",
       "      <td>2.141387</td>\n",
       "      <td>1.039230</td>\n",
       "      <td>0.930909</td>\n",
       "      <td>2.080828</td>\n",
       "      <td>1.299974</td>\n",
       "      <td>2.903225</td>\n",
       "      <td>1.291601</td>\n",
       "      <td>1.154745</td>\n",
       "      <td>...</td>\n",
       "      <td>1.670995</td>\n",
       "      <td>1.502712</td>\n",
       "      <td>2.105388</td>\n",
       "      <td>1.939682</td>\n",
       "      <td>1.180061</td>\n",
       "      <td>1.596707</td>\n",
       "      <td>1.029237</td>\n",
       "      <td>0.550024</td>\n",
       "      <td>2.121583</td>\n",
       "      <td>1.401408</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2212 rows × 583 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     3000      3001      3002      3003      3004      3006  \\\n",
       "ENSG00000000938  8.463275  8.810743  8.368607  8.466923  8.927912  9.043402   \n",
       "ENSG00000000971  1.944928  3.384911  2.522713  4.085667  3.693875  2.544924   \n",
       "ENSG00000002330  3.571424  3.959793  3.388754  4.085483  3.756835  3.986339   \n",
       "ENSG00000002549  5.872790  5.908892  6.004722  6.073372  5.541370  7.682882   \n",
       "ENSG00000002587  0.139701  1.312890  1.736479  0.976318  1.218935  0.709119   \n",
       "...                   ...       ...       ...       ...       ...       ...   \n",
       "ENSG00000271303  4.141592  5.154501  3.158052  5.093705  3.283577  5.812244   \n",
       "ENSG00000271605  3.253041  3.349505  4.194189  4.647757  3.770106  4.032242   \n",
       "ENSG00000272325  6.689842  6.335426  6.208001  6.495175  6.384606  6.810942   \n",
       "ENSG00000272886  5.031533  5.461262  4.897880  5.414562  5.042777  5.492468   \n",
       "ENSG00000277586  0.955260  1.599759  2.141387  1.039230  0.930909  2.080828   \n",
       "\n",
       "                     3007      3008      3009      3010  ...      4121  \\\n",
       "ENSG00000000938  8.681277  8.235545  8.067889  8.631065  ...  8.416943   \n",
       "ENSG00000000971  3.968554  4.085922  4.811027  4.093739  ...  2.209538   \n",
       "ENSG00000002330  3.665170  3.716128  4.060661  4.147625  ...  3.368043   \n",
       "ENSG00000002549  5.899073  5.657070  6.353882  5.949753  ...  6.898440   \n",
       "ENSG00000002587  1.053174  1.422313  2.073881  1.777967  ...  1.751712   \n",
       "...                   ...       ...       ...       ...  ...       ...   \n",
       "ENSG00000271303  4.497693  4.390843  3.943871  4.750242  ...  2.731588   \n",
       "ENSG00000271605  3.725165  4.014289  3.882068  3.648119  ...  4.301260   \n",
       "ENSG00000272325  6.646111  6.402878  6.308627  6.737337  ...  6.560920   \n",
       "ENSG00000272886  5.444711  5.371691  5.520976  5.315246  ...  5.277152   \n",
       "ENSG00000277586  1.299974  2.903225  1.291601  1.154745  ...  1.670995   \n",
       "\n",
       "                     4122      4123      4124      4125      4126      4135  \\\n",
       "ENSG00000000938  9.104202  8.838484  9.193292  8.217039  8.772340  8.497247   \n",
       "ENSG00000000971  2.390307  3.974330  2.196031  2.196303  3.910535  4.403068   \n",
       "ENSG00000002330  4.429167  3.945994  4.056662  2.965991  4.581369  4.673969   \n",
       "ENSG00000002549  6.502385  6.639824  5.425128  5.561972  6.668685  7.166978   \n",
       "ENSG00000002587  0.780928  1.562588  0.983353  1.199543  0.721738 -0.725075   \n",
       "...                   ...       ...       ...       ...       ...       ...   \n",
       "ENSG00000271303  4.644567  4.741196  4.482349  3.986768  5.829626  5.499389   \n",
       "ENSG00000271605  3.829374  4.916163  4.849648  3.918818  3.755242  4.052355   \n",
       "ENSG00000272325  6.703202  6.730188  6.621639  6.583015  6.837237  6.500973   \n",
       "ENSG00000272886  5.400093  5.523296  5.426866  5.269029  5.388664  5.499137   \n",
       "ENSG00000277586  1.502712  2.105388  1.939682  1.180061  1.596707  1.029237   \n",
       "\n",
       "                     4136      4139     41410  \n",
       "ENSG00000000938  8.612951  8.695999  8.676493  \n",
       "ENSG00000000971  2.359305  2.445675  3.267041  \n",
       "ENSG00000002330  3.146942  4.272373  3.885313  \n",
       "ENSG00000002549  6.111543  6.510586  6.325302  \n",
       "ENSG00000002587  1.494510  1.673137  0.951384  \n",
       "...                   ...       ...       ...  \n",
       "ENSG00000271303  3.893179  5.252354  4.670282  \n",
       "ENSG00000271605  3.537703  4.062805  3.979138  \n",
       "ENSG00000272325  6.102476  6.663363  6.564332  \n",
       "ENSG00000272886  4.822674  5.094705  5.293258  \n",
       "ENSG00000277586  0.550024  2.121583  1.401408  \n",
       "\n",
       "[2212 rows x 583 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49b34055",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "f763cc14",
   "metadata": {},
   "outputs": [],
   "source": [
    "X=filtered_df.T\n",
    "y=diagnosis['COHORT_DEFINITION']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "1cf5d0f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AU-ROC Score: 0.565281461833186\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.metrics import roc_auc_score\n",
    "# Split the data into training and testing sets\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=40)\n",
    "\n",
    "#  Initialize and train the XGBoost model\n",
    "xgb_model = XGBClassifier()\n",
    "xgb_model.fit(X_train, y_train)\n",
    "\n",
    "#  Make predictions on the testing set\n",
    "y_pred_proba = xgb_model.predict_proba(X_test)[:, 1]\n",
    "\n",
    "# Calculate AU-ROC score\n",
    "au_roc_score = roc_auc_score(y_test, y_pred_proba)\n",
    "\n",
    "# Print the AU-ROC score\n",
    "print(f'AU-ROC Score: {au_roc_score}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ea897e79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Column1 Data1  Column2 Data2\n",
      "0        3     C        3     X\n",
      "1        4     D        4     Y\n",
      "2        5     E        5     Z\n"
     ]
    }
   ],
   "source": [
    "# Sample DataFrames\n",
    "df1 = pd.DataFrame({'Column1': [1, 2, 3, 4, 5], 'Data1': ['A', 'B', 'C', 'D', 'E']})\n",
    "\n",
    "df2 = pd.DataFrame({'Column2': [3, 4, 5, 6, 7], 'Data2': ['X', 'Y', 'Z', 'W', 'K']})\n",
    "\n",
    "# Using merge to find the intersection\n",
    "intersection = pd.merge(df1, df2, left_on='Column1', right_on='Column2', how='inner')\n",
    "\n",
    "print(intersection)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ce838e5",
   "metadata": {},
   "source": [
    "# compare between merge and isin:\n",
    "in the merge we actually want to have the data from the both pandas dataframe files but when we use isin\n",
    "we basiccaclly want from one dataframe those that are in common with the other one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6bca5692",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column1</th>\n",
       "      <th>Data1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Column1 Data1\n",
       "0        1     A\n",
       "1        2     B\n",
       "2        3     C\n",
       "3        4     D\n",
       "4        5     E"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c4f51372",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column2</th>\n",
       "      <th>Data2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Column2 Data2\n",
       "0        3     X\n",
       "1        4     Y\n",
       "2        5     Z\n",
       "3        6     W\n",
       "4        7     K"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "raw",
   "id": "89dd7aa9",
   "metadata": {},
   "source": [
    "intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "77d130cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_values2 = df1[df1['Column1'].isin(df2['Column2'])]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5ab1b103",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column1</th>\n",
       "      <th>Data1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Column1 Data1\n",
       "2        3     C\n",
       "3        4     D\n",
       "4        5     E"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "common_values2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36f75527",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8658fcb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.reset_option('display.max_rows')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
