{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56d184cd",
   "metadata": {},
   "source": [
    "# <span style=\"color:#8B4513;\"> Machine Learning and RNA-seq Data of Parkinson Disease\n",
    "</span>\n",
    "\n",
    "\n",
    "\n",
    "[<span style=\"color:#8B4513;\">Author: **Zainab Nazari**</span>](mailto:z.nazari@ebri.com)\n",
    " \n",
    " <span style=\"color:#8B4513;\">EBRI – European Brain Research Institute Rita Levi-Montalcini | MHPC - Master in High Performance Computing</span>\n",
    " \n",
    "\n",
    "\n",
    "## Introduction\n",
    "By employing machine learning in PPMI clinical data set, we can develop predictive models that aid in the early diagnosis of the disease. These models can potentially identify specific genetic markers or gene signatures that correlate with disease progression or response to treatment.\n",
    "\n",
    "## Table of Contents\n",
    "- [Matrix of Gene Ids and Pateint Numbers](#matrixcreation)\n",
    "- [Data Preprocessing](#preprocessing)\n",
    "- [Model Training](#training)\n",
    "- [Results and Evaluation](#results)\n",
    "\n",
    "## Matrix of Gene Ids and Pateint Numbers\n",
    "- Loading the data from IR3/counts folder and extracting the associated last column (counts) of each patient file for their BL visit.\n",
    "\n",
    "\n",
    "## Data Preprocessing \n",
    "- We remove patients that have these diseases: SNCA (ENRLSNCA), GBA (ENRLGBA), LRRK2 (ENRLLRRK2).\n",
    "\n",
    "## Model Training\n",
    "Build and train machine learning models on the prepared data. Explain the choice of models, feature engineering techniques, and hyperparameter tuning. Provide code and comments to walk through the model training process.\n",
    "\n",
    "## Results and Evaluation\n",
    "Present the results of the trained models, including performance metrics, accuracy, or any relevant evaluation measures. Interpret the findings and discuss the implications. Include visualizations or tables to support the results.\n",
    "\n",
    "## Conclusion\n",
    "Summarize the key findings, limitations of the analysis, and potential future work or improvements. Offer closing remarks or suggestions for further exploration.\n",
    "\n",
    "## References\n",
    "- [**Parkinson’s Progression Markers Initiative (PPMI)**](https://www.ppmi-info.org/)\n",
    "\n",
    "- [**A Machine Learning Approach to Parkinson’s Disease Blood Transcriptomics**](https://www.mdpi.com/2073-4425/13/5/727)\n",
    "\n",
    "- [**Quality Control Metrics for Whole Blood Transcriptome Analysis in the Parkinson’s Progression Markers Initiative (PPMI)**](https://www.medrxiv.org/content/10.1101/2021.01.05.21249278v1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5c7b8ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import dask.dataframe as dd\n",
    "import os\n",
    "import glob\n",
    "import functools\n",
    "from pathlib import Path\n",
    "\n",
    "#!pip install dask[complete]; # you need to run these in case dask gives you error, it might need update.\n",
    "#!pip install --upgrade pandas \"dask[complete]\"\n",
    "#python -m pip install \"dask[dataframe]\" --upgrade"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86494e1a",
   "metadata": {},
   "source": [
    "<a id=\"matrixcreation\"></a>\n",
    "## Matrix of Gene Ids and Pateint Numbers\n",
    " Loading the data from IR3/counts folder and extracting the associated last column (counts) of each patient file for their BL visit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "403b8db8",
   "metadata": {
    "tags": [
     "matrix-creation"
    ]
   },
   "outputs": [],
   "source": [
    "# Note that the counts file in the IR3 is around 152 G, and the files are located in scratch area.\n",
    "path_to_files=\"/scratch/znazari/PPMI_ver_sep2022/RNA_Seq_data/star_ir3/counts/\"\n",
    "path1=Path(\"/scratch/znazari/PPMI_ver_sep2022/RNA_Seq_data/star_ir3/counts/\")\n",
    "\n",
    "#reading the files which are in BL (Base line) visit.\n",
    "specific_word = 'BL'\n",
    "ending_pattern = '*.txt'\n",
    "file_pattern = f'*{specific_word}*.{ending_pattern}'\n",
    "file_paths = glob.glob(path_to_files + file_pattern)\n",
    "# 'bl.txt' is a file that ccontains the name of the files with patient, BL, IR3, counts.\n",
    "filename = 'bl.txt'\n",
    "file_path_2 = os.path.join(path_to_files, filename)\n",
    "bl_files = pd.read_csv(file_path_2,header=None)\n",
    "\n",
    "# We define a function where we can take the second phrase seperated by dot. The second phrase \n",
    "# is the patient ID. So with this functin we want to get the patient IDs from their file's name\n",
    "def function_names(fname):\n",
    "    tokens=fname.split('.')\n",
    "    return tokens[1]\n",
    "\n",
    "# we create a list with the name of the each patients.\n",
    "bl_list = [function_names(bl_files.iloc[i][0]) for i in range(len(bl_files))]\n",
    "\n",
    "# here we read all the files with with base visit(BL) from the counts folder (where we have all the files\n",
    "# for all the patients and all the visit).\n",
    "list_bl_files = [dd.read_csv(path1/bl_files.iloc[i][0],skiprows=1,delimiter='\\t') for i in range(len(bl_files))]\n",
    "\n",
    "# we get the Geneid column and convert it to dask dataframe\n",
    "pd_tmp_file = list_bl_files[3].compute()\n",
    "geneid = pd_tmp_file['Geneid']\n",
    "ddf_geneid = dd.from_pandas(geneid, npartitions=1)\n",
    "\n",
    "# we get th last columns of each file in the list\n",
    "last_columns = [ddf.iloc[:, -1:] for ddf in list_bl_files]\n",
    "\n",
    "# concatinating the list of the columns in a single file.\n",
    "single_file = dd.concat(last_columns, axis=1)\n",
    "\n",
    "# we change the name of the each columns with the patient numbers.\n",
    "single_file.columns = bl_list\n",
    "\n",
    "# here we set the Geneid column as the index of the matrix.\n",
    "ddf_new_index = single_file.set_index(ddf_geneid)\n",
    "# converting to pandas data frame and saving.\n",
    "ir3_counts = ddf_new_index.compute()\n",
    "ir3_counts.to_csv(path2/\"matrix_ir3_counts_bl.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18ab10c6",
   "metadata": {},
   "source": [
    "<a id=\"preprocessing\"></a>\n",
    "## Data Preprocessing \n",
    "\n",
    "Conduct exploratory data analysis to gain insights into the dataset. Generate visualizations, compute summary statistics, and analyze relationships between variables. Include clear descriptions and captions for each plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "07749200",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10874</th>\n",
       "      <th>12499</th>\n",
       "      <th>12593</th>\n",
       "      <th>13039</th>\n",
       "      <th>13424</th>\n",
       "      <th>14281</th>\n",
       "      <th>14331</th>\n",
       "      <th>14426</th>\n",
       "      <th>15761</th>\n",
       "      <th>16580</th>\n",
       "      <th>...</th>\n",
       "      <th>60024</th>\n",
       "      <th>60036</th>\n",
       "      <th>60095</th>\n",
       "      <th>60171</th>\n",
       "      <th>65002</th>\n",
       "      <th>65003</th>\n",
       "      <th>65005</th>\n",
       "      <th>65008</th>\n",
       "      <th>70239</th>\n",
       "      <th>85242</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <td>29</td>\n",
       "      <td>18</td>\n",
       "      <td>32</td>\n",
       "      <td>11</td>\n",
       "      <td>18</td>\n",
       "      <td>20</td>\n",
       "      <td>9</td>\n",
       "      <td>15</td>\n",
       "      <td>23</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>28</td>\n",
       "      <td>6</td>\n",
       "      <td>27</td>\n",
       "      <td>19</td>\n",
       "      <td>17</td>\n",
       "      <td>44</td>\n",
       "      <td>32</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <td>877</td>\n",
       "      <td>704</td>\n",
       "      <td>574</td>\n",
       "      <td>820</td>\n",
       "      <td>677</td>\n",
       "      <td>310</td>\n",
       "      <td>863</td>\n",
       "      <td>1252</td>\n",
       "      <td>520</td>\n",
       "      <td>631</td>\n",
       "      <td>...</td>\n",
       "      <td>699</td>\n",
       "      <td>707</td>\n",
       "      <td>550</td>\n",
       "      <td>501</td>\n",
       "      <td>1049</td>\n",
       "      <td>841</td>\n",
       "      <td>1438</td>\n",
       "      <td>431</td>\n",
       "      <td>513</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <td>1686</td>\n",
       "      <td>1781</td>\n",
       "      <td>1704</td>\n",
       "      <td>1682</td>\n",
       "      <td>1656</td>\n",
       "      <td>695</td>\n",
       "      <td>1431</td>\n",
       "      <td>1778</td>\n",
       "      <td>1232</td>\n",
       "      <td>1034</td>\n",
       "      <td>...</td>\n",
       "      <td>1670</td>\n",
       "      <td>1651</td>\n",
       "      <td>1331</td>\n",
       "      <td>1013</td>\n",
       "      <td>1288</td>\n",
       "      <td>1732</td>\n",
       "      <td>2057</td>\n",
       "      <td>710</td>\n",
       "      <td>1275</td>\n",
       "      <td>858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "      <td>574</td>\n",
       "      <td>570</td>\n",
       "      <td>551</td>\n",
       "      <td>527</td>\n",
       "      <td>669</td>\n",
       "      <td>236</td>\n",
       "      <td>423</td>\n",
       "      <td>555</td>\n",
       "      <td>312</td>\n",
       "      <td>347</td>\n",
       "      <td>...</td>\n",
       "      <td>531</td>\n",
       "      <td>571</td>\n",
       "      <td>446</td>\n",
       "      <td>304</td>\n",
       "      <td>431</td>\n",
       "      <td>550</td>\n",
       "      <td>542</td>\n",
       "      <td>300</td>\n",
       "      <td>397</td>\n",
       "      <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285990.1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285991.1</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285992.1</th>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285993.1</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285994.1</th>\n",
       "      <td>237</td>\n",
       "      <td>2200</td>\n",
       "      <td>1006</td>\n",
       "      <td>633</td>\n",
       "      <td>1481</td>\n",
       "      <td>481</td>\n",
       "      <td>316</td>\n",
       "      <td>178</td>\n",
       "      <td>300</td>\n",
       "      <td>294</td>\n",
       "      <td>...</td>\n",
       "      <td>1023</td>\n",
       "      <td>1597</td>\n",
       "      <td>884</td>\n",
       "      <td>239</td>\n",
       "      <td>235</td>\n",
       "      <td>640</td>\n",
       "      <td>690</td>\n",
       "      <td>349</td>\n",
       "      <td>1253</td>\n",
       "      <td>869</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58780 rows × 1530 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    10874  12499  12593  13039  13424  14281  14331  14426  \\\n",
       "Geneid                                                                       \n",
       "ENSG00000000003.14     29     18     32     11     18     20      9     15   \n",
       "ENSG00000000005.5       3      0      9      0      0      9      0      0   \n",
       "ENSG00000000419.12    877    704    574    820    677    310    863   1252   \n",
       "ENSG00000000457.13   1686   1781   1704   1682   1656    695   1431   1778   \n",
       "ENSG00000000460.16    574    570    551    527    669    236    423    555   \n",
       "...                   ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "ENSG00000285990.1       1      2      2      0      1      0      0      0   \n",
       "ENSG00000285991.1       3      2      5      0      8      3      3      0   \n",
       "ENSG00000285992.1       7      0      5      1      1      3      0      0   \n",
       "ENSG00000285993.1       3      0      7      1      1      2      0      0   \n",
       "ENSG00000285994.1     237   2200   1006    633   1481    481    316    178   \n",
       "\n",
       "                    15761  16580  ...  60024  60036  60095  60171  65002  \\\n",
       "Geneid                            ...                                      \n",
       "ENSG00000000003.14     23      7  ...     20     28      6     27     19   \n",
       "ENSG00000000005.5       5      0  ...      0      1      0      0      2   \n",
       "ENSG00000000419.12    520    631  ...    699    707    550    501   1049   \n",
       "ENSG00000000457.13   1232   1034  ...   1670   1651   1331   1013   1288   \n",
       "ENSG00000000460.16    312    347  ...    531    571    446    304    431   \n",
       "...                   ...    ...  ...    ...    ...    ...    ...    ...   \n",
       "ENSG00000285990.1       1      0  ...      0      0      0      0      0   \n",
       "ENSG00000285991.1       1      2  ...      0      0      1      0      0   \n",
       "ENSG00000285992.1       1      0  ...      0      1      0      0      0   \n",
       "ENSG00000285993.1       3      0  ...      0      3      0      0      0   \n",
       "ENSG00000285994.1     300    294  ...   1023   1597    884    239    235   \n",
       "\n",
       "                    65003  65005  65008  70239  85242  \n",
       "Geneid                                                 \n",
       "ENSG00000000003.14     17     44     32     10      5  \n",
       "ENSG00000000005.5       2      1     14      0      0  \n",
       "ENSG00000000419.12    841   1438    431    513    314  \n",
       "ENSG00000000457.13   1732   2057    710   1275    858  \n",
       "ENSG00000000460.16    550    542    300    397    196  \n",
       "...                   ...    ...    ...    ...    ...  \n",
       "ENSG00000285990.1       1      0      0      0      0  \n",
       "ENSG00000285991.1       3      1      8      2      1  \n",
       "ENSG00000285992.1       2      0      3      1      0  \n",
       "ENSG00000285993.1       3      1      6      0      1  \n",
       "ENSG00000285994.1     640    690    349   1253    869  \n",
       "\n",
       "[58780 rows x 1530 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reading the file\n",
    "path2 = Path(\"/home/znazari/ppmi_files/data_preprocessing/data\") # where data will be saved at the end.\n",
    "read_ir3_counts = pd.read_csv(path2/\"matrix_ir3_counts_bl.csv\")\n",
    "# setting the geneid as indexing column\n",
    "read_ir3_counts.set_index('Geneid', inplace=True)\n",
    "# result\n",
    "read_ir3_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e11d6eb7",
   "metadata": {},
   "source": [
    "The patient ID needs to be as the same order as the BL matrix that we created previously."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "41ea6c9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>REC_ID</th>\n",
       "      <th>QUERY</th>\n",
       "      <th>CNO</th>\n",
       "      <th>PATNO</th>\n",
       "      <th>EVENT_ID</th>\n",
       "      <th>PAG_NAME</th>\n",
       "      <th>INFODT</th>\n",
       "      <th>LMDT</th>\n",
       "      <th>LMTM</th>\n",
       "      <th>FASTSTAT</th>\n",
       "      <th>...</th>\n",
       "      <th>PLASBFCT</th>\n",
       "      <th>DurUT1TM</th>\n",
       "      <th>DurUT1SPNTM</th>\n",
       "      <th>DurUT1FFTM</th>\n",
       "      <th>DurPLASTM</th>\n",
       "      <th>DurPLASSPNTM</th>\n",
       "      <th>DurPLASFFTM</th>\n",
       "      <th>DurBLDSERTM</th>\n",
       "      <th>DurSERSPNTM</th>\n",
       "      <th>DurSERFFTM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>272495701</td>\n",
       "      <td>NaN</td>\n",
       "      <td>001</td>\n",
       "      <td>3000</td>\n",
       "      <td>BL</td>\n",
       "      <td>LAB</td>\n",
       "      <td>02/2011</td>\n",
       "      <td>02/2011</td>\n",
       "      <td>12:00:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-194</td>\n",
       "      <td>-3</td>\n",
       "      <td>-22</td>\n",
       "      <td>-77</td>\n",
       "      <td>-15</td>\n",
       "      <td>-39</td>\n",
       "      <td>-77</td>\n",
       "      <td>-15</td>\n",
       "      <td>-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>269559601</td>\n",
       "      <td>NaN</td>\n",
       "      <td>001</td>\n",
       "      <td>3000</td>\n",
       "      <td>SC</td>\n",
       "      <td>LAB</td>\n",
       "      <td>01/2011</td>\n",
       "      <td>01/2011</td>\n",
       "      <td>12:00:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>288345201</td>\n",
       "      <td>NaN</td>\n",
       "      <td>001</td>\n",
       "      <td>3000</td>\n",
       "      <td>V01</td>\n",
       "      <td>LAB</td>\n",
       "      <td>04/2011</td>\n",
       "      <td>04/2011</td>\n",
       "      <td>07:30:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-103</td>\n",
       "      <td>-21</td>\n",
       "      <td>-65</td>\n",
       "      <td>-104</td>\n",
       "      <td>-20</td>\n",
       "      <td>-64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>308429301</td>\n",
       "      <td>NaN</td>\n",
       "      <td>001</td>\n",
       "      <td>3000</td>\n",
       "      <td>V02</td>\n",
       "      <td>LAB</td>\n",
       "      <td>08/2011</td>\n",
       "      <td>08/2011</td>\n",
       "      <td>12:00:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-190</td>\n",
       "      <td>-6</td>\n",
       "      <td>-24</td>\n",
       "      <td>-178</td>\n",
       "      <td>-35</td>\n",
       "      <td>-62</td>\n",
       "      <td>-177</td>\n",
       "      <td>-36</td>\n",
       "      <td>-63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>319985501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>001</td>\n",
       "      <td>3000</td>\n",
       "      <td>V03</td>\n",
       "      <td>LAB</td>\n",
       "      <td>11/2011</td>\n",
       "      <td>11/2011</td>\n",
       "      <td>11:45:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-98</td>\n",
       "      <td>-25</td>\n",
       "      <td>-49</td>\n",
       "      <td>-98</td>\n",
       "      <td>-25</td>\n",
       "      <td>-49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15276</th>\n",
       "      <td>3d6d0417-2d2e-4657-9fdf-cf5fd12dd227</td>\n",
       "      <td>NaN</td>\n",
       "      <td>289</td>\n",
       "      <td>160040</td>\n",
       "      <td>BL</td>\n",
       "      <td>LAB</td>\n",
       "      <td>07/2022</td>\n",
       "      <td>07/2022</td>\n",
       "      <td>09:25:00</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-55</td>\n",
       "      <td>-15</td>\n",
       "      <td>-35</td>\n",
       "      <td>0</td>\n",
       "      <td>-55</td>\n",
       "      <td>-80</td>\n",
       "      <td>0</td>\n",
       "      <td>-55</td>\n",
       "      <td>-80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15277</th>\n",
       "      <td>c90a93fa-be41-4763-9cf9-5258b45968eb</td>\n",
       "      <td>NaN</td>\n",
       "      <td>120</td>\n",
       "      <td>160231</td>\n",
       "      <td>BL</td>\n",
       "      <td>LAB</td>\n",
       "      <td>08/2022</td>\n",
       "      <td>08/2022</td>\n",
       "      <td>05:00:00</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-285</td>\n",
       "      <td>-15</td>\n",
       "      <td>-30</td>\n",
       "      <td>-290</td>\n",
       "      <td>-25</td>\n",
       "      <td>-40</td>\n",
       "      <td>-290</td>\n",
       "      <td>-25</td>\n",
       "      <td>-40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15278</th>\n",
       "      <td>0bc811a8-f682-426e-a968-bd1589d2bbcf</td>\n",
       "      <td>NaN</td>\n",
       "      <td>289</td>\n",
       "      <td>161236</td>\n",
       "      <td>BL</td>\n",
       "      <td>LAB</td>\n",
       "      <td>07/2022</td>\n",
       "      <td>07/2022</td>\n",
       "      <td>06:30:00</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-140</td>\n",
       "      <td>-85</td>\n",
       "      <td>-100</td>\n",
       "      <td>-180</td>\n",
       "      <td>-30</td>\n",
       "      <td>-60</td>\n",
       "      <td>-180</td>\n",
       "      <td>-30</td>\n",
       "      <td>-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15279</th>\n",
       "      <td>5de0cc42-b0ff-4db4-a680-f29f69b477ba</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17</td>\n",
       "      <td>162140</td>\n",
       "      <td>BL</td>\n",
       "      <td>LAB</td>\n",
       "      <td>08/2022</td>\n",
       "      <td>08/2022</td>\n",
       "      <td>20:00:00</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-850</td>\n",
       "      <td>-54</td>\n",
       "      <td>-80</td>\n",
       "      <td>-942</td>\n",
       "      <td>-40</td>\n",
       "      <td>-63</td>\n",
       "      <td>-942</td>\n",
       "      <td>-40</td>\n",
       "      <td>-63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15280</th>\n",
       "      <td>a7985659-adf9-49fd-b57a-6eebfb926406</td>\n",
       "      <td>NaN</td>\n",
       "      <td>403</td>\n",
       "      <td>162929</td>\n",
       "      <td>BL</td>\n",
       "      <td>LAB</td>\n",
       "      <td>08/2022</td>\n",
       "      <td>07/2022</td>\n",
       "      <td>21:00:00</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>-820</td>\n",
       "      <td>-25</td>\n",
       "      <td>-78</td>\n",
       "      <td>-825</td>\n",
       "      <td>-15</td>\n",
       "      <td>-33</td>\n",
       "      <td>-825</td>\n",
       "      <td>-15</td>\n",
       "      <td>-35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15281 rows × 61 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     REC_ID QUERY  CNO   PATNO EVENT_ID  \\\n",
       "0                                 272495701   NaN  001    3000       BL   \n",
       "1                                 269559601   NaN  001    3000       SC   \n",
       "2                                 288345201   NaN  001    3000      V01   \n",
       "3                                 308429301   NaN  001    3000      V02   \n",
       "4                                 319985501   NaN  001    3000      V03   \n",
       "...                                     ...   ...  ...     ...      ...   \n",
       "15276  3d6d0417-2d2e-4657-9fdf-cf5fd12dd227   NaN  289  160040       BL   \n",
       "15277  c90a93fa-be41-4763-9cf9-5258b45968eb   NaN  120  160231       BL   \n",
       "15278  0bc811a8-f682-426e-a968-bd1589d2bbcf   NaN  289  161236       BL   \n",
       "15279  5de0cc42-b0ff-4db4-a680-f29f69b477ba   NaN   17  162140       BL   \n",
       "15280  a7985659-adf9-49fd-b57a-6eebfb926406   NaN  403  162929       BL   \n",
       "\n",
       "      PAG_NAME   INFODT     LMDT      LMTM FASTSTAT  ... PLASBFCT DurUT1TM  \\\n",
       "0          LAB  02/2011  02/2011  12:00:00      NaN  ...      NaN     -194   \n",
       "1          LAB  01/2011  01/2011  12:00:00      NaN  ...      NaN      NaN   \n",
       "2          LAB  04/2011  04/2011  07:30:00      NaN  ...      NaN      NaN   \n",
       "3          LAB  08/2011  08/2011  12:00:00      NaN  ...      NaN     -190   \n",
       "4          LAB  11/2011  11/2011  11:45:00      NaN  ...      NaN      NaN   \n",
       "...        ...      ...      ...       ...      ...  ...      ...      ...   \n",
       "15276      LAB  07/2022  07/2022  09:25:00        2  ...        1      -55   \n",
       "15277      LAB  08/2022  08/2022  05:00:00        2  ...        1     -285   \n",
       "15278      LAB  07/2022  07/2022  06:30:00        2  ...        1     -140   \n",
       "15279      LAB  08/2022  08/2022  20:00:00        1  ...        1     -850   \n",
       "15280      LAB  08/2022  07/2022  21:00:00        1  ...        1     -820   \n",
       "\n",
       "      DurUT1SPNTM DurUT1FFTM DurPLASTM DurPLASSPNTM DurPLASFFTM DurBLDSERTM  \\\n",
       "0              -3        -22       -77          -15         -39         -77   \n",
       "1             NaN        NaN       NaN          NaN         NaN         NaN   \n",
       "2             NaN        NaN      -103          -21         -65        -104   \n",
       "3              -6        -24      -178          -35         -62        -177   \n",
       "4             NaN        NaN       -98          -25         -49         -98   \n",
       "...           ...        ...       ...          ...         ...         ...   \n",
       "15276         -15        -35         0          -55         -80           0   \n",
       "15277         -15        -30      -290          -25         -40        -290   \n",
       "15278         -85       -100      -180          -30         -60        -180   \n",
       "15279         -54        -80      -942          -40         -63        -942   \n",
       "15280         -25        -78      -825          -15         -33        -825   \n",
       "\n",
       "      DurSERSPNTM DurSERFFTM  \n",
       "0             -15        -39  \n",
       "1             NaN        NaN  \n",
       "2             -20        -64  \n",
       "3             -36        -63  \n",
       "4             -25        -49  \n",
       "...           ...        ...  \n",
       "15276         -55        -80  \n",
       "15277         -25        -40  \n",
       "15278         -30        -60  \n",
       "15279         -40        -63  \n",
       "15280         -15        -35  \n",
       "\n",
       "[15281 rows x 61 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path1=Path(\"/scratch/znazari/PPMI_ver_sep2022/study_data/Subject_Characteristics/\")\n",
    "mypath=Path(\"/scratch/znazari/PPMI_ver_sep2022/study_data/Lab_Collection_Procedures/\")\n",
    "cno=pd.read_csv(mypath/\"Laboratory_Procedures_with_Elapsed_Times.csv\", header=None )\n",
    "cno2=cno.rename(columns=cno.iloc[0]).drop(cno.index[0]).reset_index(drop=True)\n",
    "cno2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2d7e20b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATNO</th>\n",
       "      <th>COHORT</th>\n",
       "      <th>COHORT_DEFINITION</th>\n",
       "      <th>ENROLL_DATE</th>\n",
       "      <th>ENROLL_STATUS</th>\n",
       "      <th>STATUS_DATE</th>\n",
       "      <th>ENROLL_AGE</th>\n",
       "      <th>INEXPAGE</th>\n",
       "      <th>AV133STDY</th>\n",
       "      <th>PPMI_ONLINE_ENROLL</th>\n",
       "      <th>...</th>\n",
       "      <th>COMMENTS</th>\n",
       "      <th>CONDATE</th>\n",
       "      <th>ENRLPINK1</th>\n",
       "      <th>ENRLPRKN</th>\n",
       "      <th>ENRLSRDC</th>\n",
       "      <th>ENRLHPSM</th>\n",
       "      <th>ENRLRBD</th>\n",
       "      <th>ENRLLRRK2</th>\n",
       "      <th>ENRLSNCA</th>\n",
       "      <th>ENRLGBA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3000</td>\n",
       "      <td>2</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>02/2011</td>\n",
       "      <td>enrolled</td>\n",
       "      <td>05/2021</td>\n",
       "      <td>69.1</td>\n",
       "      <td>INEXHC</td>\n",
       "      <td>0</td>\n",
       "      <td>NO</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>06/2021</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3001</td>\n",
       "      <td>1</td>\n",
       "      <td>Parkinson's Disease</td>\n",
       "      <td>03/2011</td>\n",
       "      <td>enrolled</td>\n",
       "      <td>09/2021</td>\n",
       "      <td>65.1</td>\n",
       "      <td>INEXPD</td>\n",
       "      <td>0</td>\n",
       "      <td>NO</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>06/2021</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3002</td>\n",
       "      <td>1</td>\n",
       "      <td>Parkinson's Disease</td>\n",
       "      <td>03/2011</td>\n",
       "      <td>enrolled</td>\n",
       "      <td>09/2021</td>\n",
       "      <td>67.6</td>\n",
       "      <td>INEXPD</td>\n",
       "      <td>0</td>\n",
       "      <td>NO</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>06/2021</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3003</td>\n",
       "      <td>1</td>\n",
       "      <td>Parkinson's Disease</td>\n",
       "      <td>04/2011</td>\n",
       "      <td>enrolled</td>\n",
       "      <td>01/2022</td>\n",
       "      <td>56.7</td>\n",
       "      <td>INEXPD</td>\n",
       "      <td>0</td>\n",
       "      <td>NO</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>06/2021</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3004</td>\n",
       "      <td>2</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>04/2011</td>\n",
       "      <td>enrolled</td>\n",
       "      <td>01/2022</td>\n",
       "      <td>59.4</td>\n",
       "      <td>INEXHC</td>\n",
       "      <td>0</td>\n",
       "      <td>YES</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>06/2021</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2346</th>\n",
       "      <td>162994</td>\n",
       "      <td>1</td>\n",
       "      <td>Parkinson's Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>screened</td>\n",
       "      <td>07/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>INEXPD</td>\n",
       "      <td>0</td>\n",
       "      <td>NO</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2347</th>\n",
       "      <td>163265</td>\n",
       "      <td>1</td>\n",
       "      <td>Parkinson's Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>screened</td>\n",
       "      <td>07/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>INEXPD</td>\n",
       "      <td>0</td>\n",
       "      <td>YES</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2348</th>\n",
       "      <td>164900</td>\n",
       "      <td>1</td>\n",
       "      <td>Parkinson's Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>screened</td>\n",
       "      <td>07/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>INEXPD</td>\n",
       "      <td>0</td>\n",
       "      <td>NO</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2349</th>\n",
       "      <td>167222</td>\n",
       "      <td>1</td>\n",
       "      <td>Parkinson's Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>screened</td>\n",
       "      <td>08/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>INEXPD</td>\n",
       "      <td>0</td>\n",
       "      <td>NO</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2350</th>\n",
       "      <td>168306</td>\n",
       "      <td>1</td>\n",
       "      <td>Parkinson's Disease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>screened</td>\n",
       "      <td>08/2022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>INEXPD</td>\n",
       "      <td>0</td>\n",
       "      <td>YES</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2351 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       PATNO COHORT    COHORT_DEFINITION ENROLL_DATE ENROLL_STATUS  \\\n",
       "0       3000      2      Healthy Control     02/2011      enrolled   \n",
       "1       3001      1  Parkinson's Disease     03/2011      enrolled   \n",
       "2       3002      1  Parkinson's Disease     03/2011      enrolled   \n",
       "3       3003      1  Parkinson's Disease     04/2011      enrolled   \n",
       "4       3004      2      Healthy Control     04/2011      enrolled   \n",
       "...      ...    ...                  ...         ...           ...   \n",
       "2346  162994      1  Parkinson's Disease         NaN      screened   \n",
       "2347  163265      1  Parkinson's Disease         NaN      screened   \n",
       "2348  164900      1  Parkinson's Disease         NaN      screened   \n",
       "2349  167222      1  Parkinson's Disease         NaN      screened   \n",
       "2350  168306      1  Parkinson's Disease         NaN      screened   \n",
       "\n",
       "     STATUS_DATE ENROLL_AGE INEXPAGE AV133STDY PPMI_ONLINE_ENROLL  ...  \\\n",
       "0        05/2021       69.1   INEXHC         0                 NO  ...   \n",
       "1        09/2021       65.1   INEXPD         0                 NO  ...   \n",
       "2        09/2021       67.6   INEXPD         0                 NO  ...   \n",
       "3        01/2022       56.7   INEXPD         0                 NO  ...   \n",
       "4        01/2022       59.4   INEXHC         0                YES  ...   \n",
       "...          ...        ...      ...       ...                ...  ...   \n",
       "2346     07/2022        NaN   INEXPD         0                 NO  ...   \n",
       "2347     07/2022        NaN   INEXPD         0                YES  ...   \n",
       "2348     07/2022        NaN   INEXPD         0                 NO  ...   \n",
       "2349     08/2022        NaN   INEXPD         0                 NO  ...   \n",
       "2350     08/2022        NaN   INEXPD         0                YES  ...   \n",
       "\n",
       "     COMMENTS  CONDATE ENRLPINK1 ENRLPRKN ENRLSRDC ENRLHPSM ENRLRBD ENRLLRRK2  \\\n",
       "0         NaN  06/2021         0        0        0        0       0         0   \n",
       "1         NaN  06/2021         0        0        1        0       0         0   \n",
       "2         NaN  06/2021         0        0        1        0       0         0   \n",
       "3         NaN  06/2021         0        0        1        0       0         0   \n",
       "4         NaN  06/2021         0        0        0        0       0         0   \n",
       "...       ...      ...       ...      ...      ...      ...     ...       ...   \n",
       "2346      NaN      NaN         0        0        1        0       0         0   \n",
       "2347      NaN      NaN         0        0        1        0       0         0   \n",
       "2348      NaN      NaN         0        0        1        0       0         0   \n",
       "2349      NaN      NaN         0        0        1        0       0         0   \n",
       "2350      NaN      NaN         0        0        1        0       0         0   \n",
       "\n",
       "     ENRLSNCA ENRLGBA  \n",
       "0           0       0  \n",
       "1           0       0  \n",
       "2           0       0  \n",
       "3           0       0  \n",
       "4           0       0  \n",
       "...       ...     ...  \n",
       "2346        0       0  \n",
       "2347        0       0  \n",
       "2348        0       0  \n",
       "2349        0       0  \n",
       "2350        0       0  \n",
       "\n",
       "[2351 rows x 34 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diago=pd.read_csv(path1/\"Participant_Status.csv\", header=None )\n",
    "diago1=diago.rename(columns=diago.iloc[0]).drop(diago.index[0]).reset_index(drop=True)\n",
    "diago1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "717f7bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this is to remove patients that have these diseases: SNCA (ENRLSNCA), GBA (ENRLGBA), LRRK2 (ENRLLRRK2)\n",
    "filtered_SNCA = diago1[(diago1['ENRLSNCA'] == \"0\")& (diago1['ENRLGBA'] == \"0\")& (diago1['ENRLLRRK2'] == \"0\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d19d4d3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8b42713",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb21f2e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea67ec78",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "71f5d06c",
   "metadata": {},
   "source": [
    "<a id=\"training\"></a>\n",
    "## Model Training \n",
    "\n",
    "Build and train machine learning models on the prepared data. Explain the choice of models, feature engineering techniques, and hyperparameter tuning. Provide code and comments to walk through the model training process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80dc5964",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "27602ca3",
   "metadata": {},
   "source": [
    "<a id=\"results\"></a>\n",
    "## Results and Evaluation \n",
    "\n",
    "Present the results of the trained models, including performance metrics, accuracy, or any relevant evaluation measures. Interpret the findings and discuss the implications. Include visualizations or tables to support the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa3d5e4d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
