{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56d184cd",
   "metadata": {},
   "source": [
    " #  RNA-Seq Data Preprocessing STEP I\n",
    "\n",
    "\n",
    "-------\n",
    "\n",
    "- We keep only individuals with diagnosis of Health Control or Parkinson's Disease.\n",
    "- We remove patients that have these gene mutations : SNCA, GBA, LRRK2, and taking dopaminergic drugs.\n",
    "- We remove the duplicated gene IDs which are those that carry ensembl genes with suffix _PAR_Y and their X transcripts.\n",
    "-  We only keep genes that are either in the 19393 protein coding gene list or in 5874 long intergenic non-coding RNAs (lincRNAs) list that we obtained from the official HGNC repository (date: 31-Jan-2024).\n",
    "- We filter out genes with low expression levels, retaining only those genes that exhibit more than five counts in a minimum of 10% of the individuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5c7b8ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8ba5df20",
   "metadata": {},
   "outputs": [],
   "source": [
    "path2 = Path(\"/home/znazari/data\") # where the output data will be saved at the end.\n",
    "path3=Path(\"/scratch/znazari/PPMI_ver_sep2022/study_data/Subject_Characteristics/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "07749200",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the main table of gene IDs vs invididuals \n",
    "read_ir3_counts = pd.read_csv(path2/\"matrix_ir3_counts_bl.csv\")\n",
    "\n",
    "# Set the geneid as indexing column\n",
    "read_ir3_counts.set_index('Geneid', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "56456e0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the file which contains diagnosis\n",
    "diago=pd.read_csv(path3/\"Participant_Status.csv\", header=None )\n",
    "diago1=diago.rename(columns=diago.iloc[0]).drop(diago.index[0]).reset_index(drop=True)\n",
    "\n",
    "# Select only diagnosis with parkinson's and control.\n",
    "selected_diagnosis_pd_hc = diago1[diago1['COHORT_DEFINITION'].isin(['Healthy Control', \"Parkinson's Disease\"])]\n",
    "pd_hc = selected_diagnosis_pd_hc['PATNO']\n",
    "\n",
    "# Matrix of gene IDs with pateints only with PD and Control subjests\n",
    "filtered_df = read_ir3_counts.loc[:, read_ir3_counts.columns.isin(pd_hc)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a169c1dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the file which contains patients with gene mutations and/or using dopaminergic drugs\n",
    "union_drugs_mutations=pd.read_csv(path2/'union_drugs_mutations.csv', index_col=0)\n",
    "s_union_drugs_mutations= union_drugs_mutations['0']\n",
    "s_union_drugs_mutations_str = s_union_drugs_mutations.astype(str)\n",
    "\n",
    "# Remove patients with gene mutations and/or using dopaminergic drugs\n",
    "filtered_df_drug = filtered_df.drop(columns=s_union_drugs_mutations_str, errors='ignore')\n",
    "\n",
    "# Make some copy for later use\n",
    "filtered_df_drug_c = filtered_df_drug.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e410516",
   "metadata": {},
   "source": [
    "## Genes with  _PAR_Y and their X counterpart\n",
    "\n",
    "The suffix _PAR_Y in Ensembl ID refers to the pseudoautosomal region (PAR) of the Y chromosome (which can only be found in males). The PAR is a region of the Y chromosome that shares homology with the X chromosome, and it is the only part of the Y chromosome that recombines with the X chromosome during meiosis.\n",
    "\n",
    "For simplicity, we exclude genes associated with the Y chromosome (_PAR_Y genes), as well as their X chromosome counterparts. Alternatively  for female we could remove the _PAR_Y as it is irrelevant for them and for males we could add the value of _PAR_Y genes with their X counterparts, considering that quantification should encompass the total value of these isoforms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "79f83ac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove the version\n",
    "filtered_df_drug_c.index =filtered_df_drug.index.str.split('.').str[0]\n",
    "\n",
    "# Get the duplicated indices\n",
    "duplicated_indices = filtered_df_drug_c.index[filtered_df_drug_c.index.duplicated()]\n",
    "\n",
    "# Create a new dataframe with the duplicated indices\n",
    "new_df = filtered_df_drug_c.loc[duplicated_indices]\n",
    "\n",
    "# Sort the list of based on their indices\n",
    "new_df.sort_index(inplace=True)\n",
    "\n",
    "# Get the indices\n",
    "gene_id_duplicated = new_df.index\n",
    "\n",
    "# Make a list out of indices\n",
    "list_duplicated_genes = list(gene_id_duplicated)\n",
    "\n",
    "# Filter the list of genes with their _PAR_Y counter part and including their versions\n",
    "matching_rows = filtered_df_drug[filtered_df_drug.index.str.contains('|'.join(list_duplicated_genes), case=False)]\n",
    "\n",
    "# Make a list of genes that have duplicated Y chromosomes as well\n",
    "Genes_list_duplicated = list(matching_rows.index)\n",
    "\n",
    "filtered_df_druggg = filtered_df_drug[~filtered_df_drug.index.isin(Genes_list_duplicated)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b5e1e7e",
   "metadata": {},
   "source": [
    "## Save complete list of Gene IDs with and withOUT versions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7a35aebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine them together\n",
    "all_IR3_counts_gene_IDs = pd.DataFrame({\n",
    "    'Gene_IDs_with_version': filtered_df_drug.index,\n",
    "    'Gene_IDs_withOUT_version': filtered_df_drug_c.index})\n",
    "\n",
    "# Save complete list of Gene IDs with and withOUT versions\n",
    "all_IR3_counts_gene_IDs.to_csv(path2/\"all_IR3_counts_gene_IDs.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cc34575",
   "metadata": {},
   "source": [
    "## Ensembl gene IDs corresponding to protein-coding and long non-coding RNA genes\n",
    "\n",
    "We obtained a complete list of long non-coding RNA and protein-coding gene symbols from the official HGNC repository (date: 31-Jan-2024), that were then converted to Ensemble gene IDs using a table from BioMart (date: 31-Jan-2024). Only expression data corresponding to long non-coding RNA and protein-coding genes (24694 Ensemble gene IDs in total) were selected for further analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fc882ed4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep the genes with proteomic and non proteomic IncRNAs genes\n",
    "\n",
    "# protein coding gene symbols:\n",
    "protein_coding = pd.read_csv(\"../External_data/HGNC_protein_coding_genes_31Jan2024.txt\", delimiter='\\t')\n",
    "\n",
    "# long non-coding protein gene symbols\n",
    "non_protein_codying = pd.read_csv(\"../External_data/HGNC_lncRNA_non_coding_genes_31Jan2024.txt\", delimiter='\\t')\n",
    "\n",
    "# Dictionary to convery gene symbols to Ensemble gene IDs:\n",
    "dictionary = pd.read_csv(\"../External_data/HGNC_BioMart_symbol_name_EnsembleGeneID_31Jan2024.txt\", delimiter='\\t')\n",
    "\n",
    "# Ensemble gene IDs for protein and long non-coding protein genes:\n",
    "intersection_genes=pd.read_csv(\"../External_data/Intersection_genes_IR3_counts_gene_IDs_with_HGNC_lncRNA_Protein_Coding_04feb2024.txt\", delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2c74a2b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy the dataframe\n",
    "full_list_genes_hc_pd = filtered_df_druggg.copy() \n",
    "\n",
    "# Remove the version\n",
    "full_list_genes_hc_pd.index =full_list_genes_hc_pd.index.str.split('.').str[0]\n",
    "\n",
    "# Protein and long non-coding protein genes list\n",
    "protein_long_non_ptotein= list(intersection_genes[\"Ensemble_Gene_ID\"])\n",
    "\n",
    "# Filter the DataFrame based on the intersection of ensemble gene IDs\n",
    "full_ensemblgene = full_list_genes_hc_pd[full_list_genes_hc_pd.index.isin(protein_long_non_ptotein)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ae3dd2d",
   "metadata": {},
   "source": [
    "## Lowly expressed genes removal\n",
    "We filter out genes with low expression levels, retaining only those genes that exhibit more than five counts in a minimum of 10% of the individuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1c1cb30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the percentage of patients in which each gene has more than five counts\n",
    "gene_percentages = (full_ensemblgene > 5).sum(axis=1) / full_ensemblgene.shape[1]\n",
    "\n",
    "# Set the threshold value\n",
    "threshold = 0.1\n",
    "\n",
    "# Filter out genes whose percentage is below the threshold\n",
    "filtered_genes = gene_percentages[gene_percentages >= threshold].index\n",
    "\n",
    "# Remove the filtered genes from the dataframe\n",
    "highly_expressed_genes = full_ensemblgene.loc[filtered_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "22572a79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3000</th>\n",
       "      <th>3001</th>\n",
       "      <th>3002</th>\n",
       "      <th>3003</th>\n",
       "      <th>3004</th>\n",
       "      <th>3008</th>\n",
       "      <th>3010</th>\n",
       "      <th>3011</th>\n",
       "      <th>3012</th>\n",
       "      <th>3013</th>\n",
       "      <th>...</th>\n",
       "      <th>4075</th>\n",
       "      <th>4076</th>\n",
       "      <th>4079</th>\n",
       "      <th>4081</th>\n",
       "      <th>4091</th>\n",
       "      <th>4102</th>\n",
       "      <th>4108</th>\n",
       "      <th>4115</th>\n",
       "      <th>4136</th>\n",
       "      <th>4139</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>40</td>\n",
       "      <td>13</td>\n",
       "      <td>87</td>\n",
       "      <td>11</td>\n",
       "      <td>27</td>\n",
       "      <td>24</td>\n",
       "      <td>14</td>\n",
       "      <td>35</td>\n",
       "      <td>20</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>60</td>\n",
       "      <td>22</td>\n",
       "      <td>25</td>\n",
       "      <td>18</td>\n",
       "      <td>43</td>\n",
       "      <td>20</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>563</td>\n",
       "      <td>815</td>\n",
       "      <td>879</td>\n",
       "      <td>855</td>\n",
       "      <td>1194</td>\n",
       "      <td>980</td>\n",
       "      <td>1185</td>\n",
       "      <td>1446</td>\n",
       "      <td>672</td>\n",
       "      <td>1048</td>\n",
       "      <td>...</td>\n",
       "      <td>492</td>\n",
       "      <td>528</td>\n",
       "      <td>712</td>\n",
       "      <td>687</td>\n",
       "      <td>468</td>\n",
       "      <td>855</td>\n",
       "      <td>555</td>\n",
       "      <td>628</td>\n",
       "      <td>426</td>\n",
       "      <td>754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>1869</td>\n",
       "      <td>1510</td>\n",
       "      <td>1438</td>\n",
       "      <td>1593</td>\n",
       "      <td>2418</td>\n",
       "      <td>1607</td>\n",
       "      <td>2210</td>\n",
       "      <td>2702</td>\n",
       "      <td>1573</td>\n",
       "      <td>2573</td>\n",
       "      <td>...</td>\n",
       "      <td>923</td>\n",
       "      <td>1160</td>\n",
       "      <td>1647</td>\n",
       "      <td>1808</td>\n",
       "      <td>1223</td>\n",
       "      <td>1586</td>\n",
       "      <td>1271</td>\n",
       "      <td>1378</td>\n",
       "      <td>1037</td>\n",
       "      <td>1390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>512</td>\n",
       "      <td>367</td>\n",
       "      <td>460</td>\n",
       "      <td>444</td>\n",
       "      <td>581</td>\n",
       "      <td>488</td>\n",
       "      <td>605</td>\n",
       "      <td>922</td>\n",
       "      <td>515</td>\n",
       "      <td>834</td>\n",
       "      <td>...</td>\n",
       "      <td>291</td>\n",
       "      <td>398</td>\n",
       "      <td>540</td>\n",
       "      <td>530</td>\n",
       "      <td>356</td>\n",
       "      <td>395</td>\n",
       "      <td>438</td>\n",
       "      <td>510</td>\n",
       "      <td>343</td>\n",
       "      <td>291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285844</th>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>28</td>\n",
       "      <td>8</td>\n",
       "      <td>34</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>14</td>\n",
       "      <td>21</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>28</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>39</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>13</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285869</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285967</th>\n",
       "      <td>1043</td>\n",
       "      <td>1291</td>\n",
       "      <td>1173</td>\n",
       "      <td>963</td>\n",
       "      <td>1609</td>\n",
       "      <td>1300</td>\n",
       "      <td>1729</td>\n",
       "      <td>1950</td>\n",
       "      <td>1188</td>\n",
       "      <td>2077</td>\n",
       "      <td>...</td>\n",
       "      <td>670</td>\n",
       "      <td>776</td>\n",
       "      <td>1038</td>\n",
       "      <td>1141</td>\n",
       "      <td>784</td>\n",
       "      <td>796</td>\n",
       "      <td>890</td>\n",
       "      <td>1081</td>\n",
       "      <td>700</td>\n",
       "      <td>784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285972</th>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285985</th>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22032 rows × 545 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 3000  3001  3002  3003  3004  3008  3010  3011  3012  3013  \\\n",
       "Geneid                                                                        \n",
       "ENSG00000000003    40    13    87    11    27    24    14    35    20    35   \n",
       "ENSG00000000005     4     0    28     2    10     0     2     0     0     2   \n",
       "ENSG00000000419   563   815   879   855  1194   980  1185  1446   672  1048   \n",
       "ENSG00000000457  1869  1510  1438  1593  2418  1607  2210  2702  1573  2573   \n",
       "ENSG00000000460   512   367   460   444   581   488   605   922   515   834   \n",
       "...               ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "ENSG00000285844    23     2    28     8    34     6    11     7    14    21   \n",
       "ENSG00000285869     0     0    16     0     7     1     0     0     0     0   \n",
       "ENSG00000285967  1043  1291  1173   963  1609  1300  1729  1950  1188  2077   \n",
       "ENSG00000285972     7     2     7     1     3     0     4     1     1     4   \n",
       "ENSG00000285985     8     0     7     1    10     1     0     0     0     2   \n",
       "\n",
       "                 ...  4075  4076  4079  4081  4091  4102  4108  4115  4136  \\\n",
       "Geneid           ...                                                         \n",
       "ENSG00000000003  ...    60    22    25    18    43    20     7     8    16   \n",
       "ENSG00000000005  ...    21     1     0     0    19     0     1     0     5   \n",
       "ENSG00000000419  ...   492   528   712   687   468   855   555   628   426   \n",
       "ENSG00000000457  ...   923  1160  1647  1808  1223  1586  1271  1378  1037   \n",
       "ENSG00000000460  ...   291   398   540   530   356   395   438   510   343   \n",
       "...              ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "ENSG00000285844  ...    34    28     8     3    39     7     5     4    13   \n",
       "ENSG00000285869  ...    23     6     0     0     8     0     0     0     4   \n",
       "ENSG00000285967  ...   670   776  1038  1141   784   796   890  1081   700   \n",
       "ENSG00000285972  ...     5    10     0     0    14     2     0     0     8   \n",
       "ENSG00000285985  ...    15    14     0     0    11     0     0     1     4   \n",
       "\n",
       "                 4139  \n",
       "Geneid                 \n",
       "ENSG00000000003    15  \n",
       "ENSG00000000005     1  \n",
       "ENSG00000000419   754  \n",
       "ENSG00000000457  1390  \n",
       "ENSG00000000460   291  \n",
       "...               ...  \n",
       "ENSG00000285844    10  \n",
       "ENSG00000285869     2  \n",
       "ENSG00000285967   784  \n",
       "ENSG00000285972     2  \n",
       "ENSG00000285985     2  \n",
       "\n",
       "[22032 rows x 545 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "highly_expressed_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "53af4bc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Last update : 2024-02-06\n"
     ]
    }
   ],
   "source": [
    "# Get the current date\n",
    "current_date = datetime.now().date()\n",
    "\n",
    "# Print the current date\n",
    "print(\"Last update :\", current_date)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ac4554df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GeneID\n",
       "GeneA    0.0\n",
       "GeneB    0.5\n",
       "GeneC    1.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Assuming you have a DataFrame named df with patient IDs as columns and gene counts as rows\n",
    "# Replace these with your actual DataFrame\n",
    "\n",
    "# Sample data creation (replace this with your actual data)\n",
    "data = {'GeneID': ['GeneA', 'GeneB', 'GeneC'],\n",
    "        'Patient1': [3, 7, 10],\n",
    "        'Patient2': [1, 8, 12],\n",
    "        'Patient3': [2, 5, 9],\n",
    "       'Patient4': [2, 5, 9]}\n",
    "df = pd.DataFrame(data)\n",
    "\n",
    "# Set 'GeneID' column as the index\n",
    "df.set_index('GeneID', inplace=True)\n",
    "\n",
    "# Calculate the percentage of patients in which each gene has more than five counts\n",
    "gene_percentages = (df > 5).sum(axis=1) / df.shape[1]\n",
    "\n",
    "# Display the result\n",
    "gene_percentages\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "de88b1a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient1</th>\n",
       "      <th>Patient2</th>\n",
       "      <th>Patient3</th>\n",
       "      <th>Patient4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GeneA</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneB</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneC</th>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Patient1  Patient2  Patient3  Patient4\n",
       "GeneID                                        \n",
       "GeneA          3         1         2         2\n",
       "GeneB          7         8         5         5\n",
       "GeneC         10        12         9         9"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f5261b3d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GeneID\n",
       "GeneA    0\n",
       "GeneB    2\n",
       "GeneC    4\n",
       "dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df > 5).sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "502a2bfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient1</th>\n",
       "      <th>Patient2</th>\n",
       "      <th>Patient3</th>\n",
       "      <th>Patient4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GeneA</th>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneB</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneC</th>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Patient1  Patient2  Patient3  Patient4\n",
       "GeneID                                        \n",
       "GeneA          3         1         2         2\n",
       "GeneB          7         8         5         5\n",
       "GeneC         10        12         9         9"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3201b91b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "558bc16c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "\n",
    "# Calculate the number of counts for each gene across all patients\n",
    "#gene_counts = df.sum(axis=0)\n",
    "\n",
    "# Calculate the percentage of patients in which each gene has more than five counts\n",
    "#gene_percentages = (df > 5).sum(axis=0) / df.shape[0]\n",
    "\n",
    "# Set the threshold value\n",
    "threshold = 0.1\n",
    "\n",
    "# Filter out genes whose percentage is below the threshold\n",
    "filtered_genes = gene_percentages[gene_percentages >= threshold].index\n",
    "\n",
    "# Remove the filtered genes from the dataframe\n",
    "df_filtered = df.loc[filtered_genes]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "267bcf40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['GeneB', 'GeneC'], dtype='object', name='GeneID')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "05cd4dbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient1</th>\n",
       "      <th>Patient2</th>\n",
       "      <th>Patient3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GeneB</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneC</th>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Patient1  Patient2  Patient3\n",
       "GeneID                              \n",
       "GeneB          7         8         5\n",
       "GeneC         10        12         9"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84a81e31",
   "metadata": {},
   "outputs": [],
   "source": [
    "#rna_filtered.to_csv(path2/'ir3_rna_step1.csv', index=True)\n",
    "# combine the target as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43e0e3d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "####### here read a transcriptomic matrix, transpose it,  correlate genes with Diagnosis (0/1) using Pearson.\n",
    "#######  Select the top 10% most correlated genes in absolute values, then add a Class column as last column with PD/CTR diagnosis\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import csv\n",
    "import pandas\n",
    "import numpy\n",
    "import scipy\n",
    "\n",
    "\n",
    "working_path=\"D:\\\\disco_H\\\\bandi_grants\\\\Regione_Lazio_FILAS_2016_Confessore\\\\E_LIFE_submitted_13dic2016\\\\tesi_laurea\\\\Zainab_Nazari\\\\PPMI_analysis\\\\R_limma\"\n",
    "\n",
    "os.chdir(working_path)\n",
    "\n",
    "name_input_file=\"mydata_TMM_Norm_Log2_CPM_filtered_batch_sex_effect_removed_RIN_covariate_06feb2024.txt\"\n",
    "name_output_file_T=\"mydata_TMM_Norm_Log2_CPM_filtered_batch_sex_effect_removed_RIN_covariate_T_06feb2024.txt\"\n",
    "name_output_file_T_Class=\"mydata_TMM_Norm_Log2_CPM_filt_batch_sex_RIN_covariate_T_06feb2024_Class_PD_CTR.txt\"\n",
    "name_factor_file=\"factor_ir3_rna_step1_preprocessing_06feb2024.txt\"\n",
    "\n",
    "myfactor=pandas.read_csv(name_factor_file,  sep=\"\\t\",   header=0,index_col=0)\n",
    "myfactor2=myfactor.rename(columns={\"Diagnosis\": \"Class\"})\n",
    "myfactor2\n",
    "\n",
    "\n",
    "\n",
    "df_1=pandas.read_table(name_input_file,  header=0)\n",
    "df_1_trasp=df_1.T\n",
    "df_1_trasp_header= df_1_trasp.iloc[0,:]\n",
    "df_1_trasp=df_1_trasp.iloc[1:,]\n",
    "df_1_trasp.columns=df_1_trasp_header    #### set the 1st row of gene names as header\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# df_1_trasp.to_csv(name_output_file_T, sep=\"\\t\",header=False)  ### write and re-read to eliminate the header line\n",
    "# df_1_trasp=pandas.read_table(name_output_file_T,  header=0, index=0)\n",
    "# df_1_trasp.shape\n",
    "\n",
    "           #########here correlate genes with diagnosis 0/1\n",
    "\n",
    "\n",
    "out_Pearson_table=pandas.DataFrame(index=range(df_1_trasp.shape[1]), columns=[\"Gene\",\"Pearson_corr\",\"Absolute_Pearson_corr\",\"Pearson_pval\"])\n",
    "\n",
    "for rr in range(0,df_1_trasp.shape[1]):\n",
    "    corrM_pval=scipy.stats.pearsonr(df_1_trasp.iloc[:,rr],myfactor2[\"Diagnosis_CTR_0_PD_1\"])\n",
    "    out_Pearson_table.iloc[rr,0]=df_1_trasp.columns[rr]\n",
    "    out_Pearson_table.iloc[rr,1]=corrM_pval[0]\n",
    "    out_Pearson_table.iloc[rr,2]=abs(corrM_pval[0])\n",
    "    out_Pearson_table.iloc[rr,3]=corrM_pval[1]\n",
    "\n",
    "   \n",
    "out_Pearson_table.to_csv(\"correl_GX_to_0_1_diagnosis.txt\",sep=\"\\t\",header=True)\n",
    "\n",
    "percentile_90pc_abs_Pearson=numpy.percentile(out_Pearson_table.iloc[:,2],90.0)\n",
    "percentile_90pc_abs_Pearson\n",
    "\n",
    "bb=(out_Pearson_table[\"Absolute_Pearson_corr\"] > percentile_90pc_abs_Pearson)\n",
    "bb=list(bb)\n",
    "df_1_trasp_top_10pc_abs_Pearson=df_1_trasp.loc[:, bb]   ### take only the top 10% most correlated with diagnosis\n",
    "\n",
    "df_1_trasp_top_10pc_abs_Pearson_Class=df_1_trasp_top_10pc_abs_Pearson\n",
    "\n",
    "df_1_trasp_top_10pc_abs_Pearson_Class.insert(df_1_trasp_top_10pc_abs_Pearson.shape[1],\"Class\",list(myfactor2[\"Class\"]) )\n",
    "\n",
    "df_1_trasp_top_10pc_abs_Pearson_Class.to_csv(name_output_file_T_Class,sep=\"\\t\",header=True)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
