{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56d184cd",
   "metadata": {},
   "source": [
    " #  RNA-Seq Data Preprocessing Part I\n",
    "\n",
    "-------\n",
    "\n",
    "- We keep only individuals with diagnosis of Health Control or Parkinson's Disease.\n",
    "- We remove patients that have these gene mutations : SNCA, GBA, LRRK2, and taking dopaminergic drugs.\n",
    "- We remove the duplicated gene IDs which are those that carry ensembl genes with suffix _PAR_Y and their X transcripts.\n",
    "-  We only keep genes that are either in the 19393 protein coding gene list or in 5874 long intergenic non-coding RNAs (lincRNAs) list that we obtained from the official HGNC repository (date: 31-Jan-2024).\n",
    "- We filter out genes with low expression levels, retaining only those genes that exhibit more than five counts in a minimum of 10% of the individuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5c7b8ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8ba5df20",
   "metadata": {},
   "outputs": [],
   "source": [
    "path2 = Path(\"/home/znazari/data\") # where the output data will be saved at the end.\n",
    "path3=Path(\"/scratch/znazari/PPMI_ver_sep2022/study_data/Subject_Characteristics/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "07749200",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the main table of gene IDs vs invididuals \n",
    "read_ir3_counts = pd.read_csv(path2/\"matrix_ir3_counts_bl.csv\")\n",
    "\n",
    "# Set the geneid as indexing column\n",
    "read_ir3_counts.set_index('Geneid', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "56456e0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the file which contains diagnosis\n",
    "diago=pd.read_csv(path3/\"Participant_Status.csv\", header=None )\n",
    "diago1=diago.rename(columns=diago.iloc[0]).drop(diago.index[0]).reset_index(drop=True)\n",
    "\n",
    "# Select only diagnosis with parkinson's and control.\n",
    "selected_diagnosis_pd_hc = diago1[diago1['COHORT_DEFINITION'].isin(['Healthy Control', \"Parkinson's Disease\"])]\n",
    "pd_hc = selected_diagnosis_pd_hc['PATNO']\n",
    "\n",
    "# Matrix of gene IDs with pateints only with PD and Control subjests\n",
    "filtered_df = read_ir3_counts.loc[:, read_ir3_counts.columns.isin(pd_hc)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fe157606",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3385</th>\n",
       "      <th>3514</th>\n",
       "      <th>3390</th>\n",
       "      <th>3867</th>\n",
       "      <th>3776</th>\n",
       "      <th>3803</th>\n",
       "      <th>4058</th>\n",
       "      <th>3203</th>\n",
       "      <th>50088</th>\n",
       "      <th>3453</th>\n",
       "      <th>...</th>\n",
       "      <th>3128</th>\n",
       "      <th>3069</th>\n",
       "      <th>3965</th>\n",
       "      <th>3074</th>\n",
       "      <th>42429</th>\n",
       "      <th>3322</th>\n",
       "      <th>4025</th>\n",
       "      <th>3518</th>\n",
       "      <th>3162</th>\n",
       "      <th>3668</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>47</td>\n",
       "      <td>6</td>\n",
       "      <td>25</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>20</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <td>1818</td>\n",
       "      <td>223</td>\n",
       "      <td>777</td>\n",
       "      <td>1114</td>\n",
       "      <td>278</td>\n",
       "      <td>837</td>\n",
       "      <td>376</td>\n",
       "      <td>552</td>\n",
       "      <td>935</td>\n",
       "      <td>1096</td>\n",
       "      <td>...</td>\n",
       "      <td>478</td>\n",
       "      <td>263</td>\n",
       "      <td>191</td>\n",
       "      <td>564</td>\n",
       "      <td>737</td>\n",
       "      <td>694</td>\n",
       "      <td>1051</td>\n",
       "      <td>678</td>\n",
       "      <td>501</td>\n",
       "      <td>1687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <td>3175</td>\n",
       "      <td>472</td>\n",
       "      <td>1370</td>\n",
       "      <td>1521</td>\n",
       "      <td>692</td>\n",
       "      <td>1830</td>\n",
       "      <td>843</td>\n",
       "      <td>1601</td>\n",
       "      <td>1840</td>\n",
       "      <td>2503</td>\n",
       "      <td>...</td>\n",
       "      <td>1372</td>\n",
       "      <td>434</td>\n",
       "      <td>345</td>\n",
       "      <td>1499</td>\n",
       "      <td>2091</td>\n",
       "      <td>1657</td>\n",
       "      <td>1956</td>\n",
       "      <td>1550</td>\n",
       "      <td>1316</td>\n",
       "      <td>2324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "      <td>1080</td>\n",
       "      <td>215</td>\n",
       "      <td>417</td>\n",
       "      <td>351</td>\n",
       "      <td>202</td>\n",
       "      <td>952</td>\n",
       "      <td>344</td>\n",
       "      <td>516</td>\n",
       "      <td>441</td>\n",
       "      <td>745</td>\n",
       "      <td>...</td>\n",
       "      <td>429</td>\n",
       "      <td>124</td>\n",
       "      <td>155</td>\n",
       "      <td>582</td>\n",
       "      <td>641</td>\n",
       "      <td>581</td>\n",
       "      <td>502</td>\n",
       "      <td>586</td>\n",
       "      <td>393</td>\n",
       "      <td>837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285990.1</th>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285991.1</th>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285992.1</th>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285993.1</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285994.1</th>\n",
       "      <td>677</td>\n",
       "      <td>306</td>\n",
       "      <td>332</td>\n",
       "      <td>1150</td>\n",
       "      <td>772</td>\n",
       "      <td>720</td>\n",
       "      <td>819</td>\n",
       "      <td>1648</td>\n",
       "      <td>240</td>\n",
       "      <td>510</td>\n",
       "      <td>...</td>\n",
       "      <td>630</td>\n",
       "      <td>110</td>\n",
       "      <td>193</td>\n",
       "      <td>1478</td>\n",
       "      <td>954</td>\n",
       "      <td>573</td>\n",
       "      <td>484</td>\n",
       "      <td>474</td>\n",
       "      <td>982</td>\n",
       "      <td>227</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58780 rows × 767 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    3385  3514  3390  3867  3776  3803  4058  3203  50088  \\\n",
       "Geneid                                                                      \n",
       "ENSG00000000003.14    20    23    20    47     6    25    13    13      8   \n",
       "ENSG00000000005.5      4     6     0    12     1     3     3     1      0   \n",
       "ENSG00000000419.12  1818   223   777  1114   278   837   376   552    935   \n",
       "ENSG00000000457.13  3175   472  1370  1521   692  1830   843  1601   1840   \n",
       "ENSG00000000460.16  1080   215   417   351   202   952   344   516    441   \n",
       "...                  ...   ...   ...   ...   ...   ...   ...   ...    ...   \n",
       "ENSG00000285990.1      2     4     0     5     0     1     0     1      0   \n",
       "ENSG00000285991.1      3     8     1     4     0     0     0     0      0   \n",
       "ENSG00000285992.1      3     7     0     2     0     2     2     1      0   \n",
       "ENSG00000285993.1      3     4     4     1     0     3     4     0      0   \n",
       "ENSG00000285994.1    677   306   332  1150   772   720   819  1648    240   \n",
       "\n",
       "                    3453  ...  3128  3069  3965  3074  42429  3322  4025  \\\n",
       "Geneid                    ...                                              \n",
       "ENSG00000000003.14    19  ...    10     5    20    15     15     9    11   \n",
       "ENSG00000000005.5      1  ...     0     1    12     1      1     0     0   \n",
       "ENSG00000000419.12  1096  ...   478   263   191   564    737   694  1051   \n",
       "ENSG00000000457.13  2503  ...  1372   434   345  1499   2091  1657  1956   \n",
       "ENSG00000000460.16   745  ...   429   124   155   582    641   581   502   \n",
       "...                  ...  ...   ...   ...   ...   ...    ...   ...   ...   \n",
       "ENSG00000285990.1      0  ...     0     0     5     2      0     0     0   \n",
       "ENSG00000285991.1      0  ...     3     1     8     0      2     0     1   \n",
       "ENSG00000285992.1      2  ...     0     1     4     0      0     0     0   \n",
       "ENSG00000285993.1      0  ...     0     0     4     0      0     0     0   \n",
       "ENSG00000285994.1    510  ...   630   110   193  1478    954   573   484   \n",
       "\n",
       "                    3518  3162  3668  \n",
       "Geneid                                \n",
       "ENSG00000000003.14    13    14    34  \n",
       "ENSG00000000005.5      0     0     0  \n",
       "ENSG00000000419.12   678   501  1687  \n",
       "ENSG00000000457.13  1550  1316  2324  \n",
       "ENSG00000000460.16   586   393   837  \n",
       "...                  ...   ...   ...  \n",
       "ENSG00000285990.1      0     0     1  \n",
       "ENSG00000285991.1      2     0     5  \n",
       "ENSG00000285992.1      1     0     3  \n",
       "ENSG00000285993.1      0     0     0  \n",
       "ENSG00000285994.1    474   982   227  \n",
       "\n",
       "[58780 rows x 767 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a169c1dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the file which contains patients with gene mutations and/or using dopaminergic drugs\n",
    "union_drugs_mutations=pd.read_csv(path2/'union_drugs_mutations.csv', index_col=0)\n",
    "s_union_drugs_mutations= union_drugs_mutations['0']\n",
    "s_union_drugs_mutations_str = s_union_drugs_mutations.astype(str)\n",
    "\n",
    "# Remove patients with gene mutations and/or using dopaminergic drugs\n",
    "filtered_df_drug = filtered_df.drop(columns=s_union_drugs_mutations_str, errors='ignore')\n",
    "\n",
    "# Make some copy for later use\n",
    "filtered_df_drug_c = filtered_df_drug.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e410516",
   "metadata": {},
   "source": [
    "## Genes with  _PAR_Y and their X counterpart\n",
    "\n",
    "The suffix _PAR_Y in Ensembl ID refers to the pseudoautosomal region (PAR) of the Y chromosome (which can only be found in males). The PAR is a region of the Y chromosome that shares homology with the X chromosome, and it is the only part of the Y chromosome that recombines with the X chromosome during meiosis.\n",
    "\n",
    "For simplicity, we exclude genes associated with the Y chromosome (_PAR_Y genes), as well as their X chromosome counterparts. Alternatively  for female we could remove the _PAR_Y as it is irrelevant for them and for males we could add the value of _PAR_Y genes with their X counterparts, considering that quantification should encompass the total value of these isoforms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2c26f6e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove the version\n",
    "filtered_df_drug_c.index =filtered_df_drug.index.str.split('.').str[0]\n",
    "\n",
    "# Get the duplicated indices\n",
    "duplicated_indices = filtered_df_drug_c.index[filtered_df_drug_c.index.duplicated()]\n",
    "\n",
    "# Create a new dataframe with the duplicated indices\n",
    "new_df = filtered_df_drug_c.loc[duplicated_indices]\n",
    "\n",
    "# Sort the list of based on their indices\n",
    "new_df.sort_index(inplace=True)\n",
    "\n",
    "# Get the indices\n",
    "gene_id_duplicated = new_df.index\n",
    "\n",
    "# Make a list out of indices\n",
    "list_duplicated_genes = list(gene_id_duplicated)\n",
    "\n",
    "# Filter the list of genes with their _PAR_Y counter part and including their versions\n",
    "matching_rows = filtered_df_drug[filtered_df_drug.index.str.contains('|'.join(list_duplicated_genes), case=False)]\n",
    "\n",
    "# Make a list of genes that have duplicated Y chromosomes as well\n",
    "Genes_list_duplicated = list(matching_rows.index)\n",
    "\n",
    "filtered_df_druggg = filtered_df_drug[~filtered_df_drug.index.isin(Genes_list_duplicated)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b5e1e7e",
   "metadata": {},
   "source": [
    "## Save complete list of Gene IDs with and withOUT versions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fb903012",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine them together\n",
    "all_IR3_counts_gene_IDs = pd.DataFrame({\n",
    "    'Gene_IDs_with_version': filtered_df_drug.index,\n",
    "    'Gene_IDs_withOUT_version': filtered_df_drug_c.index})\n",
    "\n",
    "# Save complete list of Gene IDs with and withOUT versions\n",
    "all_IR3_counts_gene_IDs.to_csv(path2/\"all_IR3_counts_gene_IDs.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cc34575",
   "metadata": {},
   "source": [
    "## Ensembl gene IDs corresponding to protein-coding and long non-coding RNA genes\n",
    "\n",
    "We obtained a complete list of long non-coding RNA and protein-coding gene symbols from the official HGNC repository (date: 31-Jan-2024), that were then converted to Ensemble gene IDs using a table from BioMart (date: 31-Jan-2024). Only expression data corresponding to long non-coding RNA and protein-coding genes (24694 Ensemble gene IDs in total) were selected for further analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fc882ed4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep the genes with proteomic and non proteomic IncRNAs genes\n",
    "\n",
    "# protein coding gene symbols:\n",
    "protein_coding = pd.read_csv(\"External_data/HGNC_protein_coding_genes_31Jan2024.txt\", delimiter='\\t')\n",
    "\n",
    "# long non-coding protein gene symbols\n",
    "non_protein_codying = pd.read_csv(\"External_data/HGNC_lncRNA_non_coding_genes_31Jan2024.txt\", delimiter='\\t')\n",
    "\n",
    "# Dictionary to convery gene symbols to Ensemble gene IDs:\n",
    "dictionary = pd.read_csv(\"External_data/HGNC_BioMart_symbol_name_EnsembleGeneID_31Jan2024.txt\", delimiter='\\t')\n",
    "\n",
    "# Ensemble gene IDs for protein and long non-coding protein genes:\n",
    "intersection_genes=pd.read_csv(\"External_data/Intersection_genes_IR3_counts_gene_IDs_with_HGNC_lncRNA_Protein_Coding_04feb2024.txt\", delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a8e6cd66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy the dataframe\n",
    "full_list_genes_hc_pd = filtered_df_druggg.copy() \n",
    "\n",
    "# Remove the version\n",
    "full_list_genes_hc_pd.index =full_list_genes_hc_pd.index.str.split('.').str[0]\n",
    "\n",
    "# Protein and long non-coding protein genes list\n",
    "protein_long_non_ptotein= list(intersection_genes[\"Ensemble_Gene_ID\"])\n",
    "\n",
    "# Filter the DataFrame based on the intersection of ensemble gene IDs\n",
    "full_ensemblgene = full_list_genes_hc_pd[full_list_genes_hc_pd.index.isin(protein_long_non_ptotein)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc4a1f92",
   "metadata": {},
   "source": [
    "## Lowly expressed genes removal\n",
    "We filter out genes with low expression levels, retaining only those genes that exhibit more than five counts in a minimum of 10% of the individuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "704963aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the percentage of patients in which each gene has more than five counts\n",
    "gene_percentages = (full_ensemblgene > 5).sum(axis=1) / full_ensemblgene.shape[1]\n",
    "\n",
    "# Set the threshold value\n",
    "threshold = 0.1\n",
    "\n",
    "# Filter out genes whose percentage is below the threshold\n",
    "filtered_genes = gene_percentages[gene_percentages >= threshold].index\n",
    "\n",
    "# Remove the filtered genes from the dataframe\n",
    "highly_expressed_genes = full_ensemblgene.loc[filtered_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "84a81e31",
   "metadata": {},
   "outputs": [],
   "source": [
    "highly_expressed_genes.to_csv(path2/'ir3_rna_step1_preprocessing.csv', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "502f9d73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3385</th>\n",
       "      <th>3514</th>\n",
       "      <th>3390</th>\n",
       "      <th>3776</th>\n",
       "      <th>3803</th>\n",
       "      <th>4058</th>\n",
       "      <th>3203</th>\n",
       "      <th>3453</th>\n",
       "      <th>3633</th>\n",
       "      <th>3903</th>\n",
       "      <th>...</th>\n",
       "      <th>3332</th>\n",
       "      <th>3128</th>\n",
       "      <th>3069</th>\n",
       "      <th>3965</th>\n",
       "      <th>3074</th>\n",
       "      <th>3322</th>\n",
       "      <th>4025</th>\n",
       "      <th>3518</th>\n",
       "      <th>3162</th>\n",
       "      <th>3668</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>25</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>36</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>20</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>1818</td>\n",
       "      <td>223</td>\n",
       "      <td>777</td>\n",
       "      <td>278</td>\n",
       "      <td>837</td>\n",
       "      <td>376</td>\n",
       "      <td>552</td>\n",
       "      <td>1096</td>\n",
       "      <td>796</td>\n",
       "      <td>2136</td>\n",
       "      <td>...</td>\n",
       "      <td>508</td>\n",
       "      <td>478</td>\n",
       "      <td>263</td>\n",
       "      <td>191</td>\n",
       "      <td>564</td>\n",
       "      <td>694</td>\n",
       "      <td>1051</td>\n",
       "      <td>678</td>\n",
       "      <td>501</td>\n",
       "      <td>1687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>3175</td>\n",
       "      <td>472</td>\n",
       "      <td>1370</td>\n",
       "      <td>692</td>\n",
       "      <td>1830</td>\n",
       "      <td>843</td>\n",
       "      <td>1601</td>\n",
       "      <td>2503</td>\n",
       "      <td>1795</td>\n",
       "      <td>2225</td>\n",
       "      <td>...</td>\n",
       "      <td>1289</td>\n",
       "      <td>1372</td>\n",
       "      <td>434</td>\n",
       "      <td>345</td>\n",
       "      <td>1499</td>\n",
       "      <td>1657</td>\n",
       "      <td>1956</td>\n",
       "      <td>1550</td>\n",
       "      <td>1316</td>\n",
       "      <td>2324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>1080</td>\n",
       "      <td>215</td>\n",
       "      <td>417</td>\n",
       "      <td>202</td>\n",
       "      <td>952</td>\n",
       "      <td>344</td>\n",
       "      <td>516</td>\n",
       "      <td>745</td>\n",
       "      <td>506</td>\n",
       "      <td>493</td>\n",
       "      <td>...</td>\n",
       "      <td>461</td>\n",
       "      <td>429</td>\n",
       "      <td>124</td>\n",
       "      <td>155</td>\n",
       "      <td>582</td>\n",
       "      <td>581</td>\n",
       "      <td>502</td>\n",
       "      <td>586</td>\n",
       "      <td>393</td>\n",
       "      <td>837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285844</th>\n",
       "      <td>18</td>\n",
       "      <td>39</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>30</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>34</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285869</th>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285967</th>\n",
       "      <td>2298</td>\n",
       "      <td>337</td>\n",
       "      <td>835</td>\n",
       "      <td>478</td>\n",
       "      <td>1778</td>\n",
       "      <td>725</td>\n",
       "      <td>1098</td>\n",
       "      <td>1724</td>\n",
       "      <td>1120</td>\n",
       "      <td>1135</td>\n",
       "      <td>...</td>\n",
       "      <td>780</td>\n",
       "      <td>815</td>\n",
       "      <td>243</td>\n",
       "      <td>309</td>\n",
       "      <td>1173</td>\n",
       "      <td>1030</td>\n",
       "      <td>1298</td>\n",
       "      <td>959</td>\n",
       "      <td>898</td>\n",
       "      <td>1981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285972</th>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000285985</th>\n",
       "      <td>5</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22032 rows × 545 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 3385  3514  3390  3776  3803  4058  3203  3453  3633  3903  \\\n",
       "Geneid                                                                        \n",
       "ENSG00000000003    20    23    20     6    25    13    13    19    15    15   \n",
       "ENSG00000000005     4     6     0     1     3     3     1     1     0     0   \n",
       "ENSG00000000419  1818   223   777   278   837   376   552  1096   796  2136   \n",
       "ENSG00000000457  3175   472  1370   692  1830   843  1601  2503  1795  2225   \n",
       "ENSG00000000460  1080   215   417   202   952   344   516   745   506   493   \n",
       "...               ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "ENSG00000285844    18    39     5     9     7    12     5     7     5     6   \n",
       "ENSG00000285869     3     6     2     0     0     3     0     2     2     0   \n",
       "ENSG00000285967  2298   337   835   478  1778   725  1098  1724  1120  1135   \n",
       "ENSG00000285972     1     7     2     0     1     0     1     1     2     1   \n",
       "ENSG00000285985     5    14     2     0     0     2     2     0     0     0   \n",
       "\n",
       "                 ...  3332  3128  3069  3965  3074  3322  4025  3518  3162  \\\n",
       "Geneid           ...                                                         \n",
       "ENSG00000000003  ...    36    10     5    20    15     9    11    13    14   \n",
       "ENSG00000000005  ...     7     0     1    12     1     0     0     0     0   \n",
       "ENSG00000000419  ...   508   478   263   191   564   694  1051   678   501   \n",
       "ENSG00000000457  ...  1289  1372   434   345  1499  1657  1956  1550  1316   \n",
       "ENSG00000000460  ...   461   429   124   155   582   581   502   586   393   \n",
       "...              ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   \n",
       "ENSG00000285844  ...    30    13     5    34     5     3    13     9    12   \n",
       "ENSG00000285869  ...    12     1     0     5     0     0     0     1     0   \n",
       "ENSG00000285967  ...   780   815   243   309  1173  1030  1298   959   898   \n",
       "ENSG00000285972  ...     3     1     0     4     3     1     1     3     3   \n",
       "ENSG00000285985  ...     6     0     2     9     0     0     0     1     0   \n",
       "\n",
       "                 3668  \n",
       "Geneid                 \n",
       "ENSG00000000003    34  \n",
       "ENSG00000000005     0  \n",
       "ENSG00000000419  1687  \n",
       "ENSG00000000457  2324  \n",
       "ENSG00000000460   837  \n",
       "...               ...  \n",
       "ENSG00000285844    10  \n",
       "ENSG00000285869     2  \n",
       "ENSG00000285967  1981  \n",
       "ENSG00000285972     1  \n",
       "ENSG00000285985     0  \n",
       "\n",
       "[22032 rows x 545 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "highly_expressed_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "53af4bc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Last update : 2024-02-28\n"
     ]
    }
   ],
   "source": [
    "# Get the current date\n",
    "current_date = datetime.now().date()\n",
    "\n",
    "# Print the current date\n",
    "print(\"Last update :\", current_date)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
