{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6099c14b",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Proteomics CSF Analysis of PPMI \n",
    "We analyse **Cerebrospinal Fluid (CSF) proteomic** data of Parkinson's disease patients.\n",
    "- CSF is a clear fluid that surrounds the brain and spinal cord. It provides a protective cushion for the brain and helps remove waste products.\n",
    "- The collected CSF can be analyzed to look for various markers, proteins, or other substances that may be indicative of neurological conditions, including Parkinson's disease.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3be3a62",
   "metadata": {},
   "source": [
    "# XGboost ML from proteomic gene expressions among PD and Control patients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a53b3a87",
   "metadata": {
    "slideshow": {
     "slide_type": "skip"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import datetime\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "from sklearn.model_selection import train_test_split, cross_val_score\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.metrics import precision_score, recall_score, f1_score, roc_curve, accuracy_score, auc\n",
    "from sklearn.metrics import roc_auc_score, confusion_matrix, precision_recall_curve\n",
    "from xgboost import XGBClassifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3c33c11d",
   "metadata": {
    "slideshow": {
     "slide_type": "skip"
    }
   },
   "outputs": [],
   "source": [
    "# The pathes (Directories) to the data\n",
    "path = Path(\"/home/znazari/data/open_proteomic/\")\n",
    "path2 = Path(\"/home/znazari/data/\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77512612",
   "metadata": {},
   "source": [
    "# Excluding from the data the patients whom they are using dopaminergic drugs or have particular mutation in their genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "df6b6188",
   "metadata": {},
   "outputs": [],
   "source": [
    "union_drugs_mutations=pd.read_csv(path2/'union_drugs_mutations.csv', index_col=0)\n",
    "s_union_drugs_mutations= union_drugs_mutations['0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "161575dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_proteomic_patients=pd.read_csv(path/\"final_proteomic_patients.csv\")\n",
    "\n",
    "# Remove rows where 'PATNO' is in the values of s_union_drugs_mutations\n",
    "final_proteomic_patients_filtered = final_proteomic_patients[~final_proteomic_patients['PATNO'].isin(s_union_drugs_mutations)]\n",
    "# Reset index\n",
    "final_proteomic_patients_filtered = final_proteomic_patients_filtered.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "de2880a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>COHORT</th>\n",
       "      <th>10000-28_3</th>\n",
       "      <th>10001-7_3</th>\n",
       "      <th>10003-15_3</th>\n",
       "      <th>10006-25_3</th>\n",
       "      <th>10008-43_3</th>\n",
       "      <th>10009-2_3</th>\n",
       "      <th>10010-10_3</th>\n",
       "      <th>10011-65_3</th>\n",
       "      <th>10012-5_3</th>\n",
       "      <th>...</th>\n",
       "      <th>9984-12_3</th>\n",
       "      <th>9986-14_3</th>\n",
       "      <th>9987-30_3</th>\n",
       "      <th>9989-12_3</th>\n",
       "      <th>9991-112_3</th>\n",
       "      <th>9993-11_3</th>\n",
       "      <th>9994-217_3</th>\n",
       "      <th>9995-6_3</th>\n",
       "      <th>9997-12_3</th>\n",
       "      <th>9999-1_3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PATNO</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3029</th>\n",
       "      <td>0</td>\n",
       "      <td>6.881810</td>\n",
       "      <td>7.698871</td>\n",
       "      <td>6.235733</td>\n",
       "      <td>6.397376</td>\n",
       "      <td>5.898213</td>\n",
       "      <td>5.814846</td>\n",
       "      <td>6.204347</td>\n",
       "      <td>7.770759</td>\n",
       "      <td>6.309053</td>\n",
       "      <td>...</td>\n",
       "      <td>6.000616</td>\n",
       "      <td>9.258282</td>\n",
       "      <td>5.840534</td>\n",
       "      <td>6.320405</td>\n",
       "      <td>7.183427</td>\n",
       "      <td>6.182855</td>\n",
       "      <td>6.954768</td>\n",
       "      <td>9.604112</td>\n",
       "      <td>8.947430</td>\n",
       "      <td>9.358625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3963</th>\n",
       "      <td>1</td>\n",
       "      <td>6.843277</td>\n",
       "      <td>7.940987</td>\n",
       "      <td>6.253929</td>\n",
       "      <td>6.248572</td>\n",
       "      <td>5.881750</td>\n",
       "      <td>5.813261</td>\n",
       "      <td>6.203654</td>\n",
       "      <td>7.828046</td>\n",
       "      <td>6.263731</td>\n",
       "      <td>...</td>\n",
       "      <td>6.169101</td>\n",
       "      <td>8.977272</td>\n",
       "      <td>5.852367</td>\n",
       "      <td>6.352225</td>\n",
       "      <td>6.882810</td>\n",
       "      <td>6.252332</td>\n",
       "      <td>6.960120</td>\n",
       "      <td>9.578161</td>\n",
       "      <td>8.595830</td>\n",
       "      <td>9.151849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3316</th>\n",
       "      <td>0</td>\n",
       "      <td>6.833672</td>\n",
       "      <td>8.154299</td>\n",
       "      <td>6.226679</td>\n",
       "      <td>6.417975</td>\n",
       "      <td>5.950047</td>\n",
       "      <td>5.731352</td>\n",
       "      <td>6.329086</td>\n",
       "      <td>7.940774</td>\n",
       "      <td>6.328306</td>\n",
       "      <td>...</td>\n",
       "      <td>6.054259</td>\n",
       "      <td>8.856002</td>\n",
       "      <td>5.765146</td>\n",
       "      <td>6.259030</td>\n",
       "      <td>7.287134</td>\n",
       "      <td>6.254280</td>\n",
       "      <td>7.052670</td>\n",
       "      <td>9.688273</td>\n",
       "      <td>9.096634</td>\n",
       "      <td>9.614637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3124</th>\n",
       "      <td>1</td>\n",
       "      <td>6.685086</td>\n",
       "      <td>7.365837</td>\n",
       "      <td>6.227807</td>\n",
       "      <td>6.318911</td>\n",
       "      <td>6.109293</td>\n",
       "      <td>5.822314</td>\n",
       "      <td>6.023443</td>\n",
       "      <td>7.706822</td>\n",
       "      <td>6.288972</td>\n",
       "      <td>...</td>\n",
       "      <td>6.186214</td>\n",
       "      <td>8.766998</td>\n",
       "      <td>5.939028</td>\n",
       "      <td>6.361248</td>\n",
       "      <td>7.139187</td>\n",
       "      <td>6.364042</td>\n",
       "      <td>7.523021</td>\n",
       "      <td>8.960747</td>\n",
       "      <td>8.346206</td>\n",
       "      <td>9.197527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3175</th>\n",
       "      <td>1</td>\n",
       "      <td>6.984575</td>\n",
       "      <td>8.301633</td>\n",
       "      <td>6.211235</td>\n",
       "      <td>6.411799</td>\n",
       "      <td>5.994321</td>\n",
       "      <td>5.863646</td>\n",
       "      <td>6.282596</td>\n",
       "      <td>7.934654</td>\n",
       "      <td>6.257924</td>\n",
       "      <td>...</td>\n",
       "      <td>6.102391</td>\n",
       "      <td>9.922082</td>\n",
       "      <td>5.726641</td>\n",
       "      <td>6.259768</td>\n",
       "      <td>6.818141</td>\n",
       "      <td>6.236396</td>\n",
       "      <td>7.159039</td>\n",
       "      <td>9.780839</td>\n",
       "      <td>8.825444</td>\n",
       "      <td>9.373566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3812</th>\n",
       "      <td>0</td>\n",
       "      <td>7.014804</td>\n",
       "      <td>7.469007</td>\n",
       "      <td>6.306823</td>\n",
       "      <td>6.314757</td>\n",
       "      <td>5.893445</td>\n",
       "      <td>6.001855</td>\n",
       "      <td>6.143816</td>\n",
       "      <td>7.752982</td>\n",
       "      <td>6.460554</td>\n",
       "      <td>...</td>\n",
       "      <td>6.270356</td>\n",
       "      <td>8.784713</td>\n",
       "      <td>6.170978</td>\n",
       "      <td>6.270834</td>\n",
       "      <td>6.675186</td>\n",
       "      <td>6.919316</td>\n",
       "      <td>7.224437</td>\n",
       "      <td>8.986365</td>\n",
       "      <td>8.306326</td>\n",
       "      <td>9.158217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3817</th>\n",
       "      <td>0</td>\n",
       "      <td>6.787784</td>\n",
       "      <td>7.847977</td>\n",
       "      <td>6.257174</td>\n",
       "      <td>6.412752</td>\n",
       "      <td>5.943464</td>\n",
       "      <td>5.825788</td>\n",
       "      <td>6.183518</td>\n",
       "      <td>7.829379</td>\n",
       "      <td>6.233162</td>\n",
       "      <td>...</td>\n",
       "      <td>6.128340</td>\n",
       "      <td>9.528539</td>\n",
       "      <td>5.840993</td>\n",
       "      <td>6.326199</td>\n",
       "      <td>7.422274</td>\n",
       "      <td>6.160604</td>\n",
       "      <td>7.063162</td>\n",
       "      <td>9.438577</td>\n",
       "      <td>8.590444</td>\n",
       "      <td>9.395105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3792</th>\n",
       "      <td>1</td>\n",
       "      <td>6.881702</td>\n",
       "      <td>7.962447</td>\n",
       "      <td>6.224932</td>\n",
       "      <td>6.268376</td>\n",
       "      <td>6.019198</td>\n",
       "      <td>5.847903</td>\n",
       "      <td>6.152021</td>\n",
       "      <td>7.931321</td>\n",
       "      <td>6.283648</td>\n",
       "      <td>...</td>\n",
       "      <td>6.222008</td>\n",
       "      <td>9.231704</td>\n",
       "      <td>6.029250</td>\n",
       "      <td>6.332100</td>\n",
       "      <td>7.402976</td>\n",
       "      <td>6.506793</td>\n",
       "      <td>6.756406</td>\n",
       "      <td>9.300107</td>\n",
       "      <td>8.453467</td>\n",
       "      <td>9.147362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3793</th>\n",
       "      <td>1</td>\n",
       "      <td>6.906777</td>\n",
       "      <td>7.918118</td>\n",
       "      <td>6.138533</td>\n",
       "      <td>6.500583</td>\n",
       "      <td>5.863784</td>\n",
       "      <td>5.778317</td>\n",
       "      <td>6.155346</td>\n",
       "      <td>7.687791</td>\n",
       "      <td>6.252034</td>\n",
       "      <td>...</td>\n",
       "      <td>6.098178</td>\n",
       "      <td>9.763621</td>\n",
       "      <td>5.657412</td>\n",
       "      <td>6.234351</td>\n",
       "      <td>6.923669</td>\n",
       "      <td>6.299240</td>\n",
       "      <td>7.255767</td>\n",
       "      <td>9.800064</td>\n",
       "      <td>9.033998</td>\n",
       "      <td>9.623652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3917</th>\n",
       "      <td>0</td>\n",
       "      <td>6.926580</td>\n",
       "      <td>8.053035</td>\n",
       "      <td>6.102388</td>\n",
       "      <td>6.354829</td>\n",
       "      <td>5.987164</td>\n",
       "      <td>5.856352</td>\n",
       "      <td>6.122587</td>\n",
       "      <td>7.827367</td>\n",
       "      <td>6.221964</td>\n",
       "      <td>...</td>\n",
       "      <td>6.130150</td>\n",
       "      <td>10.000907</td>\n",
       "      <td>5.792511</td>\n",
       "      <td>6.326421</td>\n",
       "      <td>7.054380</td>\n",
       "      <td>6.570227</td>\n",
       "      <td>6.858293</td>\n",
       "      <td>9.896671</td>\n",
       "      <td>9.065101</td>\n",
       "      <td>9.351200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>555 rows × 4786 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       COHORT  10000-28_3  10001-7_3  10003-15_3  10006-25_3  10008-43_3  \\\n",
       "PATNO                                                                      \n",
       "3029        0    6.881810   7.698871    6.235733    6.397376    5.898213   \n",
       "3963        1    6.843277   7.940987    6.253929    6.248572    5.881750   \n",
       "3316        0    6.833672   8.154299    6.226679    6.417975    5.950047   \n",
       "3124        1    6.685086   7.365837    6.227807    6.318911    6.109293   \n",
       "3175        1    6.984575   8.301633    6.211235    6.411799    5.994321   \n",
       "...       ...         ...        ...         ...         ...         ...   \n",
       "3812        0    7.014804   7.469007    6.306823    6.314757    5.893445   \n",
       "3817        0    6.787784   7.847977    6.257174    6.412752    5.943464   \n",
       "3792        1    6.881702   7.962447    6.224932    6.268376    6.019198   \n",
       "3793        1    6.906777   7.918118    6.138533    6.500583    5.863784   \n",
       "3917        0    6.926580   8.053035    6.102388    6.354829    5.987164   \n",
       "\n",
       "       10009-2_3  10010-10_3  10011-65_3  10012-5_3  ...  9984-12_3  \\\n",
       "PATNO                                                ...              \n",
       "3029    5.814846    6.204347    7.770759   6.309053  ...   6.000616   \n",
       "3963    5.813261    6.203654    7.828046   6.263731  ...   6.169101   \n",
       "3316    5.731352    6.329086    7.940774   6.328306  ...   6.054259   \n",
       "3124    5.822314    6.023443    7.706822   6.288972  ...   6.186214   \n",
       "3175    5.863646    6.282596    7.934654   6.257924  ...   6.102391   \n",
       "...          ...         ...         ...        ...  ...        ...   \n",
       "3812    6.001855    6.143816    7.752982   6.460554  ...   6.270356   \n",
       "3817    5.825788    6.183518    7.829379   6.233162  ...   6.128340   \n",
       "3792    5.847903    6.152021    7.931321   6.283648  ...   6.222008   \n",
       "3793    5.778317    6.155346    7.687791   6.252034  ...   6.098178   \n",
       "3917    5.856352    6.122587    7.827367   6.221964  ...   6.130150   \n",
       "\n",
       "       9986-14_3  9987-30_3  9989-12_3  9991-112_3  9993-11_3  9994-217_3  \\\n",
       "PATNO                                                                       \n",
       "3029    9.258282   5.840534   6.320405    7.183427   6.182855    6.954768   \n",
       "3963    8.977272   5.852367   6.352225    6.882810   6.252332    6.960120   \n",
       "3316    8.856002   5.765146   6.259030    7.287134   6.254280    7.052670   \n",
       "3124    8.766998   5.939028   6.361248    7.139187   6.364042    7.523021   \n",
       "3175    9.922082   5.726641   6.259768    6.818141   6.236396    7.159039   \n",
       "...          ...        ...        ...         ...        ...         ...   \n",
       "3812    8.784713   6.170978   6.270834    6.675186   6.919316    7.224437   \n",
       "3817    9.528539   5.840993   6.326199    7.422274   6.160604    7.063162   \n",
       "3792    9.231704   6.029250   6.332100    7.402976   6.506793    6.756406   \n",
       "3793    9.763621   5.657412   6.234351    6.923669   6.299240    7.255767   \n",
       "3917   10.000907   5.792511   6.326421    7.054380   6.570227    6.858293   \n",
       "\n",
       "       9995-6_3  9997-12_3  9999-1_3  \n",
       "PATNO                                 \n",
       "3029   9.604112   8.947430  9.358625  \n",
       "3963   9.578161   8.595830  9.151849  \n",
       "3316   9.688273   9.096634  9.614637  \n",
       "3124   8.960747   8.346206  9.197527  \n",
       "3175   9.780839   8.825444  9.373566  \n",
       "...         ...        ...       ...  \n",
       "3812   8.986365   8.306326  9.158217  \n",
       "3817   9.438577   8.590444  9.395105  \n",
       "3792   9.300107   8.453467  9.147362  \n",
       "3793   9.800064   9.033998  9.623652  \n",
       "3917   9.896671   9.065101  9.351200  \n",
       "\n",
       "[555 rows x 4786 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save the final list of patients and their proteomic csf data into file\n",
    "final_proteomic_patients_filtered.to_csv(path/'final_proteomic_patients_filtered.csv', index=False)\n",
    "\n",
    "# Set the 'Name' column as the index\n",
    "final_proteomic_patients_filtered.set_index('PATNO', inplace=True)\n",
    "\n",
    "df = final_proteomic_patients_filtered\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f6d1c0d",
   "metadata": {},
   "source": [
    "## Pearson's Correlation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "01beadfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extracting gene expression and diagnosis columns\n",
    "gene_expression = df.drop('COHORT', axis=1)  # Drop 'Diagnosis' column from gene expression data\n",
    "diagnosis = df['COHORT']\n",
    "\n",
    "# Calculate Pearson correlation for each gene and store in a dictionary\n",
    "correlation_dict = {}\n",
    "for gene in gene_expression.columns:\n",
    "    correlation_coefficient, _ = pearsonr(gene_expression[gene], diagnosis)\n",
    "    correlation_dict[gene] = abs(correlation_coefficient)\n",
    "\n",
    "# Sort the genes by the absolute correlation coefficient in descending order\n",
    "sorted_genes = sorted(correlation_dict.items(), key=lambda x: x[1], reverse=True)\n",
    "\n",
    "# Extract the top 50% of genes\n",
    "top_20_percent = int(0.5 * len(sorted_genes))\n",
    "selected_genes = [gene[0] for gene in sorted_genes[:top_20_percent]]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bba2c7c7",
   "metadata": {},
   "source": [
    "## XGboost algorithm with hyperparametrs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d91da4c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AU-ROC Score: 0.7889\n"
     ]
    }
   ],
   "source": [
    "# Define hyperparameters and seed\n",
    "hyperparameters = {'learning_rate': 0.1, 'max_depth': 1, 'n_estimators': 300}\n",
    "seed = 42\n",
    "\n",
    "# Keep only the columns in the selected_genes list\n",
    "selected_gene_expression = gene_expression[selected_genes]\n",
    "\n",
    "# Separate features (X) and target variable (y)\n",
    "X = selected_gene_expression\n",
    "y = diagnosis\n",
    "\n",
    "# Split the data into training and testing sets\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=41)\n",
    "\n",
    "# Feature scaling\n",
    "scaler = StandardScaler()\n",
    "X_train_scaled = scaler.fit_transform(X_train)\n",
    "X_test_scaled = scaler.transform(X_test)\n",
    "\n",
    "# Initialize and train the XGBoost model\n",
    "xgb_model = XGBClassifier(**hyperparameters, seed=seed)\n",
    "xgb_model.fit(X_train_scaled, y_train)\n",
    "\n",
    "# Make predictions on the testing set\n",
    "y_pred_proba = xgb_model.predict_proba(X_test_scaled)[:, 1]\n",
    "\n",
    "# Calculate AU-ROC score\n",
    "au_roc_score = roc_auc_score(y_test, y_pred_proba)\n",
    "print(f'AU-ROC Score: {au_roc_score:.4f}')\n",
    "\n",
    "# Optional: Cross-validation\n",
    "cv_scores = cross_val_score(xgb_model, X_train_scaled, y_train, cv=10, scoring='roc_auc')\n",
    "print(f'Cross-validated AU-ROC Scores: {np.mean(cv_scores):.4f} (±{np.std(cv_scores):.4f})')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f38ebf5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the false positive rate (FPR) and true positive rate (TPR)\n",
    "fpr, tpr, _ = roc_curve(y_test, y_pred_proba)\n",
    "\n",
    "# Calculate the Area Under the ROC Curve (AUC)\n",
    "roc_auc = auc(fpr, tpr)\n",
    "\n",
    "# Create a new figure\n",
    "plt.figure(figsize=(6, 5))\n",
    "\n",
    "# Plot the ROC curve\n",
    "plt.plot(fpr, tpr, color='tab:purple', lw=2, label=f'AU-ROC Curve (AUC = {roc_auc:.2f})')\n",
    "plt.plot([0, 1], [0, 1], color='tab:orange', lw=2, linestyle='--')\n",
    "\n",
    "# Customize plot\n",
    "plt.xlabel('False Positive Rate')\n",
    "plt.ylabel('True Positive Rate')\n",
    "plt.title('Receiver Operating Characteristic (ROC) Curve')\n",
    "plt.grid(True, linestyle='--', alpha=0.5)\n",
    "plt.legend(loc='lower right')\n",
    "\n",
    "# Show the plot\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65939a7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the current date\n",
    "current_date = datetime.now().date()\n",
    "\n",
    "# Print the current date\n",
    "print(\"Last update :\", current_date)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Slideshow",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
